<!DOCTYPE HTML>
<html lang="en" class="light" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>chipseq_workflow - St. Jude Cloud RFCs</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body class="sidebar-visible no-js">
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('light')
            html.classList.add(theme);
            var body = document.querySelector('body');
            body.classList.remove('no-js')
            body.classList.add('js');
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var body = document.querySelector('body');
            var sidebar = null;
            var sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            body.classList.remove('sidebar-visible');
            body.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="introduction.html">Introduction</a></li><li class="chapter-item expanded "><a href="0001-rnaseq-workflow-v2.0.0.html"><strong aria-hidden="true">1.</strong> 0001-rnaseq-workflow-v2.0.0</a></li><li class="chapter-item expanded "><a href="0002-scRNA-Seq-workflow.html"><strong aria-hidden="true">2.</strong> 0002-scRNA-Seq-workflow</a></li><li class="chapter-item expanded "><a href="chipseq_workflow.html" class="active"><strong aria-hidden="true">3.</strong> chipseq_workflow</a></li><li class="chapter-item expanded affix "><a href="drafts.html">RFC Drafts</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">St. Jude Cloud RFCs</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="table-of-contents"><a class="header" href="#table-of-contents">Table of Contents <!-- omit in toc --></a></h1>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#motivation">Motivation</a></li>
<li><a href="#discussion">Discussion</a></li>
<li><a href="#specification">Specification</a></li>
</ul>
<h1 id="introduction"><a class="header" href="#introduction">Introduction</a></h1>
<p>This RFC lays out the specification for the ChIP-Seq mapping pipeline.</p>
<h1 id="motivation"><a class="header" href="#motivation">Motivation</a></h1>
<p>To provide the epigenetics community access to data from ChIP-Seq experiments performed at St. Jude Children's Research Hospital, we propose the following data harmonization pipeline. The goal of this pipeline is to provide harmonized alignments for ChIP-Seq data. For this pipeline, we will make no recommendations on downstream analysis, focusing instead on harmonizing the underlying sequencing data and leaving analysis decisions to the user.</p>
<h1 id="discussion"><a class="header" href="#discussion">Discussion</a></h1>
<h2 id="aligner-choice"><a class="header" href="#aligner-choice">Aligner choice</a></h2>
<p>The epigenetics community and the results analyses of ChIP-seq experiments are highly varied. Unlike RNA-Seq or DNA-Seq, there is no standard mapping method for the analysis of human data. The community primarily uses either <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a>, <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie 2</a>, or <a href="http://bio-bwa.sourceforge.net/">BWA</a> for alignment. These aligners use similar internal data structures for mapping via an FM-index.</p>
<p>To provide harmonized ChIP-Seq data in St. Jude Cloud, we need to select a single alignment method. Therefore we investigated how alignment was handled by other large projects.</p>
<h3 id="open-source"><a class="header" href="#open-source">Open source</a></h3>
<p>Open source projects are ones that publish the source code of their analysis pipelines. This allows anyone to freely explore and review the details of each step.</p>
<div class="table-wrapper"><table><thead><tr><th>Project</th><th>Stable release</th><th>Aligner</th></tr></thead><tbody>
<tr><td><a href="https://github.com/ENCODE-DCC/chip-seq-pipeline">ENCODE-DCC/chip-seq-pipeline</a> [<a href="http://doi.org/10.1101/gr.136184.111">1</a>]</td><td>1.4 (2018-03-08)</td><td>BWA 0.7.10 (2014-07-14) (bwa-aln)</td></tr>
<tr><td><a href="https://github.com/ENCODE-DCC/chip-seq-pipeline2">ENCODE-DCC/chip-seq-pipeline2</a></td><td>1.8.0 (2021-03-26)</td><td>Bowtie 2 2.3.4.3 (2018-09-17) (default)<br />BWA 0.7.17 (2017-11-07) (If read-length &gt;= 70 bp, bwa-mem; otherwise, bwa-aln.)</td></tr>
<tr><td><a href="https://nf-co.re/chipseq">nf-core/chipseq</a> [<a href="https://www.nature.com/articles/s41587-020-0439-x">1</a>]</td><td>1.2.1 (2020-07-29)</td><td>BWA 0.7.17 (2017-11-07) (bwa-mem)</td></tr>
<tr><td><a href="https://github.com/stjude/seaseq">stjude/SEASEQ</a></td><td>1.1 (2021-04-19)</td><td>Bowtie 1.2.3 (2019-07-05)</td></tr>
</tbody></table>
</div>
<h3 id="description-only"><a class="header" href="#description-only">Description only</a></h3>
<p>Some projects only provide written descriptions of their analysis pipelines, either published in a paper or some other online document. These projects do not make their source code publicly available or easily accessible.</p>
<div class="table-wrapper"><table><thead><tr><th>Project</th><th>Publication year</th><th>Aligner</th></tr></thead><tbody>
<tr><td><a href="https://chip-atlas.org/">ChIP-Atlas</a> [<a href="https://doi.org/10.15252/embr.201846255">1</a>, <a href="https://github.com/inutano/chip-atlas/wiki#2-primary-processing">2</a>]</td><td>2018</td><td>Bowtie 2 2.2.2 (2014-04-10)</td></tr>
<tr><td><a href="http://summit.med.unideb.hu/summitdb/">ChIPSummitDB</a> [<a href="https://doi.org/10.1093/database/baz141">1</a>]</td><td>2020</td><td>BWA</td></tr>
<tr><td><a href="http://cistrome.org/">Cistrome Project</a> [<a href="https://doi.org/10.1093/nar/gky1094">1</a>]</td><td>2018</td><td>BWA</td></tr>
<tr><td><a href="http://www.roadmapepigenomics.org/">Roadmap Epigenomics Project</a> [<a href="https://doi.org/10.1038/nature14248">1</a>]</td><td>2015</td><td>PASH 3.0 (2010)</td></tr>
</tbody></table>
</div>
<p>Within St. Jude, Computational Biology has historically used BWA to map ChIP-Seq data. Also, the St. Jude Center for Applied Bioinformatics uses BWA as their standard aligner for ChIP-Seq experiments.</p>
<h2 id="multiple-mapped-reads"><a class="header" href="#multiple-mapped-reads">Multiple mapped reads</a></h2>
<p>To reduce false positives, ChIP-Seq experiments commonly filter out reads that are not uniquely mapped.  In order to simplify the ChIP-Seq workflow and to provide the most widely usable alignment data, however, we do not filter reads.</p>
<h1 id="specification"><a class="header" href="#specification">Specification</a></h1>
<p>With the publication of Brian Abraham's <a href="https://doi.org/10.1186/s12859-022-04588-z">SEAseq paper</a>, the St. Jude Cloud team has opted to incorporate their ChIP-Seq pipeline as the core alignment method for St. Jude Cloud harmonized ChIP-Seq data. The <a href="https://github.com/stjude/seaseq">SEAseq repository</a> provides additional details on their method.</p>
<h2 id="required-metadata"><a class="header" href="#required-metadata">Required Metadata</a></h2>
<p>Due to the nature of the experimental methods, ChIP-Seq requires additional metadata to be understood and used for analysis. The following information must be provided for all ChIP-Seq samples to enable their release on St. Jude Cloud.</p>
<ul>
<li>Antibody manufacturer/vendor</li>
<li>Antibody product/catalog number</li>
<li>Antibody lot number</li>
<li>Antibody target/mark</li>
</ul>
<h2 id="dependencies"><a class="header" href="#dependencies">Dependencies</a></h2>
<p>Dependencies are managed via Docker images. Where possible, community maintained (Biocontainers) images are utilized. These will be updated over time.</p>
<ul>
<li><a href="https://github.com/broadinstitute/picard">picard</a></li>
<li><a href="http://www.htslib.org/">samtools</a></li>
<li><a href="https://github.com/deeptools/deepTools">deeptools</a></li>
<li><a href="https://github.com/stjude-rust-labs/fq">fq</a></li>
<li><a href="https://github.com/stjude/seaseq">SEAseq</a></li>
</ul>
<h2 id="reference-files"><a class="header" href="#reference-files">Reference files</a></h2>
<p>The following reference files are used as the basis of the ChIP-Seq Workflow:</p>
<ul>
<li>
<p>Similarly to all analysis pipelines in St. Jude Cloud, we use the <code>GRCh38_no_alt</code> analysis set for our reference genome. You can get a copy of the file <a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz">here</a>. Additionally, you can get the file by running the following commands:</p>
<pre><code class="language-bash">wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz -O GRCh38_no_alt.fa.gz
gunzip GRCh38_no_alt.fa.gz

echo &quot;a6da8681616c05eb542f1d91606a7b2f  GRCh38_no_alt.fa&quot; &gt; GRCh38_no_alt.fa.md5
md5sum -c GRCh38_no_alt.fa.md5
# &gt; GRCh38_no_alt.fa: OK
</code></pre>
</li>
<li>
<p>The bowtie reference file is preparted using SEAseq.</p>
<pre><code class="language-bash"> bowtie-build \
  --threads $ncpu \    # Number of threads for building index
  $reference_file \    # FASTA file to use for index
  $output_index
</code></pre>
</li>
<li>
<p>For the gene model, we use the GENCODE v31 &quot;comprehensive gene annotation&quot; GTF for the &quot;CHR&quot; regions. You can get a copy of the gene annotation file <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gtf.gz">here</a>. For the exact steps to generate the gene model we use, you can run the following commands:</p>
<pre><code class="language-bash">wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gtf.gz
echo &quot;0aac86e8a6c9e5fa5afe2a286325da81  gencode.v31.annotation.gtf.gz&quot; &gt; gencode.v31.annotation.gtf.gz.md5
md5sum -c gencode.v31.annotation.gtf.gz.md5
# &gt; gencode.v31.annotation.gtf.gz: OK

gunzip gencode.v31.annotation.gtf.gz
echo &quot;4e22351ae216e72aa57cd6d6011960f8  gencode.v31.annotation.gtf&quot; &gt; gencode.v31.annotation.gtf.md5
md5sum -c gencode.v31.annotation.gtf.md5
# &gt; gencode.v31.annotation.gtf: OK
</code></pre>
</li>
<li>
<p>A list of excluded regions are obtained from the Boyle lab. This list is based on ENCODE data and a comprehensive set of anomalous, unstructured, or high-signal regions in next generation sequencing data.</p>
<pre><code class="language-bash">wget -O hg38.excludelist.bed.gz https://github.com/Boyle-Lab/Blacklist/raw/master/lists/hg38-blacklist.v2.bed.gz
echo &quot;83fe6bf8187a64dee8079b80f75ba289  hg38.excludelist.bed.gz&quot; &gt; hg38.excludelist.bed.gz.md5
md5sum -c hg38.excludelist.bed.gz.md5
# &gt; hg38.excludelist.bed.gz: OK

gunzip hg38.excludelist.bed.gz
echo &quot;766848c2a632d1b9dd12fac98ce02bd3  hg38.excludelist.bed&quot; &gt; hg38.excludelist.bed.md5
md5sum -c hg38.excludelist.bed.md5
# &gt; hg38.excludelist.bed: OK
</code></pre>
</li>
</ul>
<h2 id="workflow"><a class="header" href="#workflow">Workflow</a></h2>
<p>Here are the resulting steps in the ChIP-Seq Workflow pipeline. There might be slight alterations in the actual implementation, which can be found in <a href="https://github.com/stjudecloud/workflows/blob/main/workflows/chipseq/chipseq-standard.wdl">the St. Jude Cloud workflows repository</a>.</p>
<ol>
<li>
<p>Run <code>picard ValidateSam</code> on the incoming BAM to ensure that it is well-formed enough to convert back to FastQ.</p>
<pre><code class="language-bash">picard ValidateSamFile I=$INPUT_BAM \                # Input BAM.
                  IGNORE=INVALID_PLATFORM_VALUE \    # Validations to ignore.
                  IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Run <code>samtools quickcheck</code> on incoming BAM to ensure it is properly formatted.</p>
<pre><code class="language-bash">samtools quickcheck $INPUT_BAM
</code></pre>
</li>
<li>
<p>Split BAM file into multiple BAMs on the different read groups using <code>samtools split</code>. See <a href="http://www.htslib.org/doc/samtools.html">the samtools documentation</a> for more information.</p>
<pre><code class="language-bash">samtools split -u $UNACCOUNTED_BAM_NAME \ # Reads that do not belong to a read group or the read group is unrecognized go here.
               -f '%*_%!.%.'              # Format of output BAM file names.
</code></pre>
<p>If the BAM has unaccounted reads, those reads will need to be removed and the samtools split step will need to be rerun.</p>
</li>
<li>
<p>Run samtools <code>collate</code> and <code>fastq</code> on each of the BAMs generated in the previous step.</p>
<pre><code class="language-bash">   if [ &quot;~{paired_end}&quot; == 'true' ]
   then
      samtools collate \
          INPUT_BAM \
      | samtools fastq
          -1 sample_R1.fastq.gz
          -2 sample_R2.fastq.gz
   else
      samtools collate \
          INPUT_BAM \
      | samtools fastq
          -1 sample.fastq.gz
   fi
</code></pre>
</li>
<li>
<p>Run <code>fq lint</code> on each of the FastQ pairs generated in the previous step as a quality check. You can see the checks that the <code>fq</code> tool performs <a href="https://github.com/stjude/fqlib/blob/master/README.md#validators">here</a>.</p>
<pre><code class="language-bash">if [ &quot;~{paired_end}&quot; == 'true' ]
then
   fq lint $FASTQ_R1 $FASTQ_R2            # Files for read 1 and read 2.
else
   fq lint $FASTQ_R1                      # Files for read 1.
fi
</code></pre>
</li>
<li>
<p>Run the <code>SEAseq</code> alignment algorithm. The following is the SEAseq method written out. We will use the implementation at https://github.com/stjude/seaseq/blob/master/workflows/workflows/mapping.wdl. </p>
<pre><code class="language-bash">        if [ -f &quot;$fastqfile_R2&quot; ]; then    # If paired-end reads
         bowtie \
             --chunkmbs=256 \                # Max MB of RAM for best-first search frames
             -p $ncpu \                      # Number of threads to use
             -k $good_alignments \           # report N good alignments per read (default: 2)
             -m $limit_alignments \          # Supress all alignments if &gt; N alignments exist (default: 2)
             -X $insert_size \               # Maximum insert size for paired-end reads (default: 600)
             $stranded_m \                   # Strandedness of reads (fr, rf, ff)
             --best \                        # Hits guaranteed best stratum
             -S \                            # write SAM format output
             $BOWTIE_INDEXES \               # Bowtie index files
             -1 $fastqfile \                 # Mate 1
             -2 $fastqfile_R2 \              # Mate 2
             &gt; $outputfile                   # Output file name
     else                                    # If single-end reads
         bowtie \
             -l $readlength \                # seed length for -n (default: input read length or 75)
             -p $ncpu \                      # Number of threads to use
             -k $good_alignments \           # report N good alignments per read (default: 2)
             -m $limit_alignments \          # Supress all alignments if &gt; N alignments exist (default: 2)
             --best \                        # Hits guaranteed best stratum
             -S \                            # write SAM format output
             $BOWTIE_INDEXES \               # Bowtie index files
             $fastqfile \                    # Fastq file containing reads
             &gt; $outputfile                   # Output file name
     fi

      if [ &quot;~{paired_end}&quot; == 'true' ]; then
         awk -F\\t 'BEGIN{j=0}{j++}{if(NF&gt;5 &amp;&amp; j%2==0){ \
             printf &quot;%s_%.0f\t&quot;, $1, j-1 } else if(NF&gt;5 &amp;&amp; j%2==1){ \
             printf &quot;%s_%.0f\t&quot;, $1, j } else { printf $1 &quot;\t&quot;;j=0 } \
             for(i=2;i&lt;=NF;i++){ printf &quot;%s\t&quot;, $i}; printf &quot;\n&quot; }' \
             $samfile &gt; $renamed_sam

         samtools fixmate -m \
             $renamed_sam \
             $fixmatefile
         samtools sort \
             $fixmatefile \
             -o $outputfile
     else
         samtools sort \
             $samfile \
             -o $outputfile
     fi

     samtools flagstat $bamfile &gt; $flagstat

     samtools index $bamfile

     # Filter BAM if excluded list provided
     intersectBed \
         -v \
         -a $bamfile \
         -b $EXCLUDE_FILE \
         &gt; $FILTERED_BAM
      # If paired:
      pairToBed \
         -abam $fixmatefile \
         -b $EXCLUDE_FILE \
         -type neither \
         &gt; $PAIRTOBED_OUTPUT_FILE

      samtools sort \
         $PAIRTOBED_OUTPUT_FILE \
         -o $PAIRTOBED_FILE

      samtools flagstat $PAIRTOBED_FILE &gt; $PAIRTOBED_FILE_FLAGSTAT

      samtools index $PAIRTOBED_FILE

      samtools markdup \
         -r -s \
         $PAIRTOBED_FILE \
         $exclude_markdup

      samtools flagstat $exclude_markdup &gt; $exclude_markdup_flagstat

      samtools index $exclude_markdup


</code></pre>
</li>
<li>
<p>Run <code>picard CleanSam</code> on the <code>SEAseq</code>-aligned BAM file. Fixes soft-clipping beyond the end-of-reference and sets MAPQ to 0 for unmapped reads.</p>
<pre><code class="language-bash">picard -Xmx${JAVA_HEAP_SIZE}g CleanSam \
   I=${BAM} \                                      # Input BAM.
   O=${OUTPUT_FILENAME}                            # Output cleaned BAM
</code></pre>
</li>
<li>
<p>Run <code>picard ValidateSamFile</code> on the aligned and sorted BAM file.</p>
<pre><code class="language-bash">picard ValidateSamFile I=$SEAseq_SORTED_BAM \         # SEAseq-aligned coordinate-sorted BAM.
               IGNORE=INVALID_PLATFORM_VALUE \        # Validations to ignore.
               IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Run <code>bamCoverage</code> to generate bigwig file.</p>
<pre><code class="language-bash">bamCoverage --bam ${MERGED_BAM} \                 # Input BAM file
            --outFileName ${PREFIX}.bw \          # Output bigwig filename
            --outFileFormat bigwig \              # Set output format to bigwig
            --numberOfProcessors &quot;max&quot;            # Utilize all available processors
</code></pre>
</li>
</ol>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="0002-scRNA-Seq-workflow.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="drafts.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="0002-scRNA-Seq-workflow.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="drafts.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
