<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>0002-quality-check-workflow - St. Jude Cloud RFCs</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="introduction.html">Introduction</a></li><li class="chapter-item expanded "><a href="0001-rnaseq-workflow-v2.0.0.html"><strong aria-hidden="true">1.</strong> 0001-rnaseq-workflow-v2.0.0</a></li><li class="chapter-item expanded "><a href="0002-quality-check-workflow.html" class="active"><strong aria-hidden="true">2.</strong> 0002-quality-check-workflow</a></li><li class="chapter-item expanded "><a href="0002-scRNA-Seq-workflow.html"><strong aria-hidden="true">3.</strong> 0002-scRNA-Seq-workflow</a></li><li class="chapter-item expanded affix "><a href="drafts.html">RFC Drafts</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">St. Jude Cloud RFCs</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="table-of-contents"><a class="header" href="#table-of-contents">Table of Contents <!-- omit in toc --></a></h1>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#motivation">Motivation</a></li>
<li><a href="#discussion">Discussion</a></li>
<li><a href="#specification">Specification</a></li>
<li><a href="#items-still-in-progress">Items Still In-Progress</a></li>
<li><a href="#outstanding-questions">Outstanding Questions</a></li>
</ul>
<h2 id="introduction"><a class="header" href="#introduction">Introduction</a></h2>
<p>This RFC documents an automated workflow for assessing the integrity and quality of St. Jude Cloud genomics data. The goal of this RFC is two-fold.</p>
<ol>
<li>Establish state of the art method for comprehensively evaluating genomics data quality at scale, both at time of receipt and after processing.</li>
<li>Publish a collection of metrics that end-users of St. Jude Cloud can leverage to assess the quality of the data available. This context should save users time computing the information themselves while also informing appropriate use of the data.</li>
</ol>
<p>You can find the relevant discussion on the <a href="https://github.com/stjudecloud/rfcs/pull/3">associated pull request</a>.</p>
<h2 id="motivation"><a class="header" href="#motivation">Motivation</a></h2>
<p>St. Jude Cloud is a large repository of omics data available for request from
the academic and non-profit community. As such, the project processes thousands
of samples from whole genome, whole exome, RNA-Seq, and various omics-based
assays each year. A standard, robust method to assess pre-processing and
post-processing quality for samples has been developed in-house, but there are
some shortcomings with our current approach. In particular, this RFC will:</p>
<ul>
<li>Define the standard set of QC tools used to evaluating omics-based data,</li>
<li>identify and implement key metrics that can be automated to assist in manual observation of the data, and</li>
<li>outline mechanisms to publish those QC results alongside the data already in
St. Jude Cloud so that end-users can leverage this information.</li>
</ul>
<h2 id="discussion"><a class="header" href="#discussion">Discussion</a></h2>
<h3 id="types-of-qc"><a class="header" href="#types-of-qc">Types of QC</a></h3>
<p>There are (at least) two different types of QC typically carried out omics-based data. <strong>Experimental</strong> QC attempts to identify the success of the assay(s) performed. Once one is sufficiently satisfied that the data generated by an experiment is &quot;good&quot;, <strong>computational</strong> QC examines the degree to which computational processing of that data was completed successfully.</p>
<p>By the time data reaches the St. Jude Cloud team from various sources, extensive
<em>experimental</em> and <em>computational</em> evaluation have already been carried out.
Each contributing project has its own thresholds for quality in both areas,
which is dependent on the best practices at that point in time and the goals of
the project. Most often, our team takes the computational data, reverts it back
to its raw form (such as FastQ files), and reprocesses the data using a
harmonization pipeline.</p>
<p>Thus, the scope of this RFC, and the QC of samples on the project in general, is
limited to the <em>computational</em> QC of the files produced for publication in St.
Jude Cloud. While we do produce results that define <em>experimental</em> results (such
as <code>fastqc</code>), these are rarely used to decide which files pass or fail our
QC—this is in recognition of the fact that the data, while not always perfect,
is extremely valuable due to its relative scarcity. We hope that the inclusion
of these results will save end-users time and aid in decision-making about
downstream analysis approaches.</p>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th><th>ChIP-Seq</th></tr></thead><tbody>
<tr><td>M Reads Mapped (<a href="#m-reads-mapped">link</a>)</td><td><a href="http://www.htslib.org/doc/samtools.html">samtools</a></td><td>Number of reads mapped in millions. This metric is useful</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads Aligned (<a href="#percentage-of-reads-aligned">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Number of mapped reads divided by the total number of reads as a percentage.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>Median Insert Size (<a href="#median-insert-size">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Median size of the fragment that is inserted between the sequencing adapters (estimated in silico).</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>Percentage Duplication (<a href="#percentage-duplication">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Percentage of the reads that are marked as PCR or optical duplicates.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>≥ 30X Coverage (<a href="#-30x-coverage">link</a>)</td><td><a href="https://github.com/brentp/mosdepth">mosdepth</a></td><td>The percentage of locations that are covered by at least 30 reads for the whole genome, the exonic regions, and the coding sequence regions specifically.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>Predicted Instrument (<a href="#predicted-instrument">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted sequencing machine that produced this data. This should match your expectations for what machine sequenced the data.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Predicted Read Length (<a href="#predicted-read-length">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted read length for the sequencing experiment. This should match your expectations for how the library was prepared. A value of -1 indicates that a stable read length could not be predicted.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Probable Encoding (<a href="#probable-encoding">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted encoding of the <em>original</em> FASTQ file. For this, all modern data should match <code>Sanger/Illumina 1.8</code>.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads attributed to Homo sapiens (<a href="#percentage-of-reads-attributed-to-homo-sapiens">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were assigned as originating from Homo sapiens from <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads attributed to top 5 species (<a href="#percentage-of-reads-attributed-to-top-5-species">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were assigned to the top 5 reported species from <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads Unclassified (<a href="#percentage-of-reads-unclassified">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were not classified by <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-green" alt="Yes" /></td></tr>
</tbody></table>
</div>
<h2 id="general-statistics"><a class="header" href="#general-statistics">General statistics</a></h2>
<h3 id="m-reads-mapped"><a class="header" href="#m-reads-mapped"><strong>M Reads Mapped</strong></a></h3>
<p><em>Produced by <a href="http://www.htslib.org/doc/samtools.html">samtools</a></em></p>
<p>Number of reads mapped in millions. The number of reads in a whole genome
sequencing file can vary widely based on the experiment design and the
sequencing depth of the project.</p>
<ul>
<li>Typically, WGS files contain between 200M reads up to 2 billion reads (or
higher in some cases). Any deviation outside of this range should be
investigated further (particularly if the number is lower, which suggests
that the file may be truncated).</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect a normal distribution for this metric within the same
sequencing cohort, special attention is required for any samples that do not
conform to this distribution. </li>
</ul>
<h3 id="percentage-of-reads-aligned"><a class="header" href="#percentage-of-reads-aligned"><strong>Percentage of Reads Aligned</strong></a></h3>
<p><em>Produced by <a href="https://broadinstitute.github.io/picard/">picard</a></em></p>
<p>Percentage of the number of reads that were able to be mapped to a given
reference genome. Within St. Jude Cloud, this metric is indicative of the
amount of non-human material which was sequenced during the experiment.</p>
<ul>
<li>This number is especially informative within the context of samples from the
same cohort. We expect the percentage of aligned reads to be tightly
clustered above 95% mapping rate. It is not uncommon for samples to go as
low as 80% mapping rate, and this should be considered within an acceptable
range. Occasionally you may find small proportion of samples between 51%-79%
mapping rate: those samples should be investigated, but typically, you
should still include those samples in the release. Anything below 50%
mapping rate signals an issue with contamination, and anything less than 30%
mapping may be an issue with the mapper or the data integrity (e.g.
mismatched read pairs).</li>
</ul>
<h3 id="median-insert-size"><a class="header" href="#median-insert-size"><strong>Median Insert Size</strong></a></h3>
<p><em>Produced by <a href="https://broadinstitute.github.io/picard/">picard</a></em></p>
<p>Size of the fragment inserted between the sequencing adapters. In whole genome
sequencing experiments, typically you have a target fragment length for each
sample (commonly ~2x the read length to maximize the amount of information
gathered per fragment). The median insert size has proven to be an incredibly
valuable statistic for identifying mapping problems and experimental
abnormalities.</p>
<ul>
<li>Typically, the median insert size should be between 1.5x - 2x the read
length of the experiment. This criteria may be evaluated loosely, as insert
sizes that are close to this range (anything between 1x - 5x read length) do
not necessarily indicate a major problem. Samples with an extremely low
median insert size (10-40 nucleotides) are likely the cause of mapping
artifacts—particularly if <code>bwa mem</code> was used with a low seed size.</li>
<li>Insert size distributions tend to follow a <a href="https://en.wikipedia.org/wiki/Gamma_distribution">gamma
distribution</a> where long,
rolling peaks are interspersed with tight peaks close to the target insert
size.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect the median insert size across samples to be tightly
clustered within a cohort.</li>
</ul>
<h3 id="-30x-coverage"><a class="header" href="#-30x-coverage"><strong>≥ 30X Coverage</strong></a></h3>
<p><em>Produced by <a href="https://github.com/brentp/mosdepth">mosdepth</a></em></p>
<p>Percentage of the genome which has at least 30 reads covering each position (on
average) for the whole genome, the exonic regions, and the coding regions
(&quot;30x coverage&quot;). 30x coverage at any location is widely considered to be the
minimum number of reads needed to generate high confidence genotype call. In
whole genome sequencing, it is desirable to have as much of the genome as
possible to be covered by 30 or more reads. Expected genome coverage is often
determined at the time of sequencing and may vary from project to project
depending on cost and experimental design.</p>
<ul>
<li>Target sequencing depth and fraction of the genome to be covered are
generally set on a project by project basis. Speak with the someone on the
sequencing team to ensure you understand what sequencing depth and fraction
of the genome covered you should be expecting.</li>
<li>At a minimum, we expect 80% of the whole genome to be covered by at least 30
reads in high-quality whole genome sequencing experiments. For higher depth
coverage projects like our Clinical Genomics project, the standard is 60x
across 80% of the genome.</li>
<li>Less coverage does not necessarily indicate a problem, but in whole genome
sequencing, anything below 65% of the genome at 30X signals that something
may be going wrong. Pay close attention to the distribution of the reads
across the genome by generating a coverage plot to see where sequencing bias
is being introduced.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect this metric to be tightly clustered across samples
within a cohort.</li>
</ul>
<h3 id="gc-content"><a class="header" href="#gc-content"><strong>GC Content</strong></a></h3>
<p><em>Produced by <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></em></p>
<p>Percentage of the nucleotides contained within reads which are Cs (Cytosine) or
Gs (Guanine). GC content is important because Gs pair with Cs in DNA with 3
hydrogen bonds whereas As pair with Ts in DNA with 2 hydrogen bonds. Ergo, DNA
is considered to be stronger when comprised of more GC bonds, and GC content is
a defining characteristic of different genomes. GC content for humans is
<a href="https://www.nature.com/articles/35057062#Sec15">estimated to be at around 41%</a>.</p>
<ul>
<li>GC content in whole genome sequencing is typically between 40%-60% depending
on sequencing bias. Any deviation outside of this range should be
investigated further.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect GC content across samples to be tightly clustered
within a cohort.</li>
<li>If either of the above assumptions do not hold true, probable sources of
variance include library preparation abnormalities or contamination issues.</li>
</ul>
<h2 id="specification"><a class="header" href="#specification">Specification</a></h2>
<p>These are generic instructions for running each of the tools in our pipeline. We run our pipeline as a <a href="https://github.com/stjudecloud/workflows/blob/master/workflows/qc/quality-check-standard.wdl">WDL workflow</a>. We have supplied examples of the commands used for each package. For the typical memory requirements of each command, please see our <a href="https://github.com/stjudecloud/workflows">WDL repository</a>.</p>
<h3 id="dependencies"><a class="header" href="#dependencies">Dependencies</a></h3>
<p>We presume Anaconda is available and installed. If not, please follow the link to <a href="https://www.anaconda.com/">Anaconda</a> first.</p>
<pre><code class="language-bash">conda create --name bio-qc \
    --channel bioconda \
    --channel conda-forge \
    fastqc==0.11.8 \
    picard==2.20.2  \
    qualimap==2.2.2c \
    samtools==1.9 \
    fastq-screen==0.13.0 \
    ngsderive==2.2.0 \
    multiqc==1.10.1 \
    -y

conda activate bio-qc
</code></pre>
<p>For linting created fastqs, <code>fqlib</code> must be installed. See installation instructions <a href="https://github.com/stjude/fqlib">here</a>.</p>
<h3 id="workflow"><a class="header" href="#workflow">Workflow</a></h3>
<p>The workflow specification is as follows. Note that arguments that are not integral to the command (such as output directories) or vary between compute environments (such as memory thresholds or number of threads) are not included.</p>
<ol>
<li>
<p>Compute the md5 checksum of the file.</p>
<pre><code class="language-bash">md5sum $BAM
</code></pre>
</li>
<li>
<p>Use Picard's <code>ValidateSamFile</code> tool to ensure the inner contents of the BAM file are well-formed.</p>
<pre><code class="language-bash">picard ValidateSamFile \
    I=$BAM             \  # specify bam file
    MODE=SUMMARY          # concise output
</code></pre>
</li>
<li>
<p>Run <code>samtools flagstat</code> to gather general statistics such as alignment percentage.</p>
<pre><code class="language-bash">samtools flagstat $BAM
</code></pre>
</li>
<li>
<p>Run <code>fastqc</code> to collect sequencing and library-related statistics. These are only for informational purposes -- as stated above, we typically do not remove samples based on this information (with rare exception), as the sequencing-related QC work was done upstream in the genomics lab.</p>
<pre><code class="language-bash">fastqc $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive instrument</code> to infer sequencing instrument.</p>
<pre><code class="language-bash">ngsderive instrument $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive readlen</code> to infer read lengths.</p>
<pre><code class="language-bash">ngsderive readlen $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive encoding</code> to infer PHRED score encoding.</p>
<pre><code class="language-bash">ngsderive encoding \
    -n -1          \  # parse the entire file
    $BAM
</code></pre>
</li>
<li>
<p>Run <code>qualimap bamqc</code> to gather more in-depth statistics about read stats, coverage, mapping quality, mismatches, etc.</p>
<pre><code class="language-bash">qualimap bamqc -bam $BAM \  # bam filename
    -nt $NUM_THREADS     \  # threads requested
    -nw 400                 # number of windows
</code></pre>
</li>
<li>
<p>If WGS or WES data, run <code>fastq_screen</code>. For performance, we subsample the input BAM using <code>samtools view -s $computed_fraction</code> before running it through <code>picard SamToFastq</code>. The resulting fastqs are validated with <code>fq lint</code> provided by <code>fqlib</code>.</p>
<pre><code class="language-bash">cat $fastq_1 $fastq_2 &gt; $combined_fastq
fastq_screen          \
    --aligner bowtie2 \
    $combined_fastq
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>ngsderive strandedness</code> to determine a backwards-computed strandedness of the RNA-Seq experiment.</p>
<pre><code class="language-bash">ngsderive strandedness $BAM
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>ngsderive junction-annotation</code> to calculate the number of known, novel, and partial-novel junctions.</p>
<pre><code class="language-bash">ngsderive junction-annotation $BAM
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>qualimap rnaseq</code> to gather QC statistics that are tailored for RNA-Seq files.</p>
<pre><code class="language-bash">qualimap rnaseq --java-mem-size=$MEM_SIZE \  # memory
   -bam $BAM                              \  # bam filename
   [-pe]                                     # specify paired end if paired end
</code></pre>
</li>
<li>
<p>Combine all of the above metrics using <code>multiqc</code>.</p>
<pre><code class="language-bash">multiqc . # recurse all files in '.'
</code></pre>
</li>
</ol>
<h2 id="items-still-in-progress"><a class="header" href="#items-still-in-progress">Items Still In-Progress</a></h2>
<ul>
<li><input disabled="" type="checkbox"/>
Analysis tools for other types of sequencing (ChIP-seq)</li>
<li><input disabled="" type="checkbox"/>
Useful metadata from various stages (sample collection, laboratory, pre-sequencing, sequencing, post-sequencing)</li>
</ul>
<h2 id="outstanding-questions"><a class="header" href="#outstanding-questions">Outstanding Questions</a></h2>
<ul>
<li>What thresholds or metrics differentiate a poor-quality sample from a high-quality one?</li>
<li>What other metrics or properties would be valuable?</li>
<li>What is the best way to define and handle outliers?</li>
<li>What is the best way to examine cohort integrity? This means experimental category-based tests of samples to find outliers that are of sufficient quality if examined alone. Outliers in this case may indicate classification errors or rare biological conditions. Which metrics are best tested here?</li>
</ul>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="0001-rnaseq-workflow-v2.0.0.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="0002-scRNA-Seq-workflow.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="0001-rnaseq-workflow-v2.0.0.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="0002-scRNA-Seq-workflow.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
