<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>St. Jude Cloud RFCs</title>
        <meta name="robots" content="noindex" />


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="introduction.html">Introduction</a></li><li class="chapter-item expanded "><a href="0001-rnaseq-workflow-v2.0.0.html"><strong aria-hidden="true">1.</strong> 0001-rnaseq-workflow-v2.0.0</a></li><li class="chapter-item expanded "><a href="0002-quality-check-workflow.html"><strong aria-hidden="true">2.</strong> 0002-quality-check-workflow</a></li><li class="chapter-item expanded "><a href="0002-scRNA-Seq-workflow.html"><strong aria-hidden="true">3.</strong> 0002-scRNA-Seq-workflow</a></li><li class="chapter-item expanded affix "><a href="drafts.html">RFC Drafts</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">St. Jude Cloud RFCs</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <p align="center">
  <a href="https://github.com/stjudecloud/rfcs"><img src="https://github.com/stjudecloud/rfcs/raw/master/docs/rfcs-banner-blueprint.jpg" width="800" title="St. Jude Cloud RFCs"></a>
  <a href="https://actions-badge.atrox.dev/stjudecloud/rfcs/goto"><img alt="Build Status" src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fstjudecloud%2Frfcs%2Fbadge&style=flat" /></a>
  <a href="https://gitter.im/stjudecloud/rfcs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge" target="_blank">
    <img alt="Gitter: RFCs" src="https://badges.gitter.im/stjudecloud/rfcs.svg" />
  </a>
</p>
<blockquote>
<p>This repository contains all Request for Comments (or <a href="https://en.wikipedia.org/wiki/Request_for_Comments">rfcs</a>) for the St. Jude Cloud project. Currently, RFCs on the St. Jude Cloud project are focused on outlining changes to the genomic analysis pipelines. However, we may expand to other areas over time.</p>
</blockquote>
<h3 id="-homepage"><a class="header" href="#-homepage">üè† <a href="https://stjudecloud.github.io/rfcs/">Homepage</a></a></h3>
<h2 id="process-and-evaluation"><a class="header" href="#process-and-evaluation">Process and Evaluation</a></h2>
<p>Initiation of an RFC for the St. Jude Cloud project is currently limited to members of the St. Jude Cloud team. If you are a member of our community and have an idea you would like us to consider (or if you would like to propose your own RFC), please <a href="mailto:support@stjude.cloud">contact us</a> before investing a large amount of time in your proposal.</p>
<h3 id="overview"><a class="header" href="#overview">Overview</a></h3>
<p>All RFCs go through a three-stage process from draft to adoption:</p>
<ol>
<li>An initial draft is constructed by the author and discussed internally with members of the core St. Jude Cloud team. 
<ul>
<li>Titles of pull requests in this phase are typical prefixed with <code>[WIP]</code> and should not be considered mature enough to accept comments from the community. </li>
<li>The amount of time the RFC spends in this state is variable depending on the scope of the proposal.</li>
</ul>
</li>
<li>The RFC opens for discussion from users and stakeholders<sup><a href="introduction.html#footnote1">1</a></sup> within the St. Jude community. 
<ul>
<li>The beginning of this phase will be indicated by a comment on the pull request from the author.</li>
<li>The RFC will remain in this phase for <strong>1 week</strong>.</li>
</ul>
</li>
<li>The RFC opens for discussion from the broader community.
<ul>
<li>The beginning of this phase will be indicated by a comment on the pull request from the author.</li>
<li>The RFC will remain in this phase for <strong>2 weeks</strong>.</li>
</ul>
</li>
</ol>
<p>Each phase of the process brings about further refinement of details and hardening of the proposal.</p>
<p><sup><a name="footnote1">1</a></sup> St. Jude community stakeholders include the Department of Computational Biology, Center for Applied Bioinformatics, and users of St. Jude Cloud that have requested data from a <code>@stjude.org</code> e-mail address.</p>
<h3 id="evaluation-and-acceptance"><a class="header" href="#evaluation-and-acceptance">Evaluation and acceptance</a></h3>
<p>RFCs must have majority support within the St. Jude Cloud team before being adopted. In practice, most RFCs already have a significant amount of internal support before being drafted, so the rejection of an RFC will be exceedingly rare. The intent of these discussions is mostly geared towards gaining feedback from the community. </p>
<p>We welcome all ideas and concerns as we seek to understand how the needs of the genomics community evolve over time. Our ultimate goal is to make informed decisions that will benefit the majority of our users. That said, the project's direction is not driven strictly by community consensus and ultimately will be decided by the St. Jude Cloud team based on our best judgement. </p>
<h3 id="creating-a-proposal"><a class="header" href="#creating-a-proposal">Creating a proposal</a></h3>
<ol>
<li>Create an initial draft of the RFC by creating a pull request on this repo.
<ul>
<li>Create a branch on this repo with the form <code>&lt;username&gt;/&lt;featurename&gt;</code>.</li>
<li>On that branch, copy the <code>0000-template.md</code> file to the <code>text/</code> folder and replace &quot;template&quot; with an appropriate feature name (e.g. <code>text/rnaseq-workflow-v2</code>).</li>
<li>If necessary, you can create a folder at <code>resources/&lt;filename.md&gt;/</code> to hold any images or supporting materials (e.g. <code>resources/0000-rnaseq-workflow-v2.0/</code>) </li>
<li>Ensure that your pull request has <code>[WIP]</code> prefixing the title as long as it is not ready for review!</li>
</ul>
</li>
<li>Open up the discussion for the community.
<ul>
<li>Ensure your pull request's main comment has a &quot;rendered&quot; tag for easy review (see <a href="https://github.com/stjudecloud/rfcs/pull/1">this example</a>).</li>
<li>Remove the <code>[WIP]</code> if necessary.</li>
<li>Assign yourself as the assignee and add any relevant community members you'd like to review.</li>
</ul>
</li>
<li>Once the RFC has reached a consensus, it is ready for inclusion in the main repository.
<ul>
<li>An owner of this repository will comment on the pull request with an assigned RFC number.</li>
<li>Change your filename and resources folder to reflect this change (e.g. rename <code>text/0000-rnaseq-workflow-v2.0.md</code> to <code>text/1234-rnaseq-workflow-v2.0.md</code>).</li>
<li>Ensure all of your links still work in your rendered copy.</li>
<li>Merge in the PR and delete the branch.</li>
</ul>
</li>
</ol>
<h2 id="install"><a class="header" href="#install">Install</a></h2>
<pre><code class="language-sh">cargo install mdbook
</code></pre>
<h2 id="usage"><a class="header" href="#usage">Usage</a></h2>
<pre><code class="language-sh">mdbook build
python3 -m http.server -d book
# visit the rendered version in your browser at http://localhost:8000.
</code></pre>
<h2 id="author"><a class="header" href="#author">Author</a></h2>
<p>üë§ <strong>St. Jude Cloud Team</strong></p>
<ul>
<li>Website: https://stjude.cloud</li>
<li>Github: <a href="https://github.com/stjudecloud">@stjudecloud</a></li>
<li>Twitter: <a href="https://twitter.com/StJudeResearch">@StJudeResearch</a></li>
</ul>
<h2 id="contributing"><a class="header" href="#contributing">Contributing</a></h2>
<p>Contributions, issues and feature requests are welcome!<br />Feel free to check <a href="https://github.com/stjudecloud/rfcs/issues">issues page</a>. You can also take a look at the <a href="https://github.com/stjudecloud/rfcs/blob/master/CONTRIBUTING.md">contributing guide</a>.</p>
<h2 id="-license"><a class="header" href="#-license">üìù License</a></h2>
<p>Copyright ¬© 2020 <a href="https://github.com/stjudecloud">St. Jude Cloud Team</a>.<br />
This project is <a href="https://github.com/stjudecloud/rfcs/blob/master/LICENSE.md">MIT</a> licensed.</p>
<h2 id="questions"><a class="header" href="#questions">Questions</a></h2>
<p>With any questions, please <a href="mailto:support@stjude.cloud">contact us</a>.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="table-of-contents"><a class="header" href="#table-of-contents">Table of Contents <!-- omit in toc --></a></h1>
<ul>
<li><a href="0001-rnaseq-workflow-v2.0.0.html#introduction">Introduction</a></li>
<li><a href="0001-rnaseq-workflow-v2.0.0.html#motivation">Motivation</a></li>
<li><a href="0001-rnaseq-workflow-v2.0.0.html#discussion">Discussion</a></li>
<li><a href="0001-rnaseq-workflow-v2.0.0.html#specification">Specification</a></li>
<li><a href="0001-rnaseq-workflow-v2.0.0.html#appendix">Appendix</a></li>
</ul>
<h1 id="introduction"><a class="header" href="#introduction">Introduction</a></h1>
<p>This RFC lays out the specification for the RNA-Seq mapping pipeline v2.0.0. The improvements contained within are largely based on (a) new version of tools/reference files and (b) feedback from the community. You can find the relevant discussion on the <a href="https://github.com/stjudecloud/rfcs/pull/1">associated pull request</a>.</p>
<h1 id="motivation"><a class="header" href="#motivation">Motivation</a></h1>
<ul>
<li><strong>Tool additions and updates.</strong> The tools used in the first version of the RNA-seq pipeline are now woefully out of date (~2 years old). We desire to upgrade the tools to their latest published versions available so we can enjoy the benefits they provide. Furthermore, we'd like to explore adding many more tools to begin publishing QC results (probably a different QC pipeline in the future). See the <a href="0001-rnaseq-workflow-v2.0.0.html#tool-additions-updates">section below on tool additions and updates</a> for more details.</li>
<li><strong>Updated reference files.</strong> No changes have really been made to the <code>GRCh38_no_alt</code> analysis set reference genome. However, three major releases of the GENCODE gene model have transpired since we released the first revision of the RNA-Seq workflow. <a href="https://www.gencodegenes.org/human/release_31.html">GENCODE v31</a> is the latest at the time of writing, so we'd like to utilize the new information contained there as well.</li>
<li><strong>Quality of life improvements based on feedback from the community.</strong> Input from the community has greatly informed our thoughts on what can be improved in this release. Some ideas that come to mind are:
<ul>
<li>Removal of the <code>ERCC SpikeIn</code> sequences to ensure consistent sequence dictionaries between DNA and RNA data.
<ul>
<li>Popular tools such as <code>GATK</code> and <code>picard</code> are generally unhappy if the sequence dictionaries don't match perfectly.</li>
<li>The inclusion of the External RNA Controls Consortium (ERCC) Spike-in Control RNA sequences in the alignment reference file we used for RNA-Seq mapping was hence causing issues when using mapped RNA-Seq BAM files in conjunction with other non-RNA-Seq BAM files in downstream analysis using these tools.</li>
<li>Last, many of our RNA-Seq samples were not generated using 'ERCC' spike-in control sequences, so the benefit its inclusion provides is minimal.</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1 id="discussion"><a class="header" href="#discussion">Discussion</a></h1>
<h2 id="tool-additions-and-upgrades"><a class="header" href="#tool-additions-and-upgrades">Tool additions and upgrades</a></h2>
<p>As part of the RNA-Seq workflow v2.0.0, multiple tools will be added and upgraded:</p>
<ul>
<li><code>fq v0.2.0</code> (<a href="https://github.com/stjude/fqlib/releases/tag/v0.2.0">Released</a> November 28, 2018) will be added. This tool will be used to validate the output of <code>picard SamToFastq</code>. <code>picard SamToFastq</code> does not currently catch all of the errors we wish to catch at this stage (such as duplicate read names in the FastQ file). Thus, we will leverage this tool to independently validate that the data is well-formed by our definition of that phrase.</li>
<li><code>ngsderive v1.0.2</code> (<a href="https://github.com/claymcleod/ngsderive/releases/tag/v1.0.2">Released</a> March 3, 2020) will be added. This tool will be used to empirically infer strandedness of RNA-seq experiments.</li>
<li>Update <code>STAR 2.5.3a</code> (<a href="https://github.com/alexdobin/STAR/releases/tag/2.5.3a">Released</a> March 17, 2017) to <code>STAR 2.7.1a</code> (<a href="https://github.com/alexdobin/STAR/releases/tag/2.7.1a">Released</a> May 15, 2019). Upgraded to receive the benefits of bug fixes and software optimizations.</li>
<li>Update <code>samtools 1.4.0</code> (<a href="https://github.com/samtools/samtools/releases/tag/1.4">Released</a> March 13, 2017) to <code>samtools 1.9</code> (<a href="https://github.com/samtools/samtools/releases/tag/1.9">Released</a> July 18, 2018). Updating the samtools version whenever possible is of particular interest to me due to the historical fragility of the samtools code (although it has seemed to get better over the last year or so).</li>
<li>Update <code>picard 2.9.4</code> (<a href="https://github.com/broadinstitute/picard/releases/tag/2.9.4">Released</a> June 15, 2017) to <code>picard 2.20.2</code> (<a href="https://github.com/broadinstitute/picard/releases/tag/2.20.2">Released</a> May 28, 2019). Upgraded to receive the benefits of bug fixes and software optimizations.</li>
</ul>
<h2 id="gencode-compatability"><a class="header" href="#gencode-compatability">GENCODE compatability</a></h2>
<p>One of the major discussions during this round of revisions was the compatability of the <code>GRCh38_no_alt</code> reference genome with the latest <code>GENCODE</code> gene set. It was posed as the following question:</p>
<blockquote>
<p>This has just been an outstanding question of mine for a while ‚Äî how big of an impact (if any at all) does the mismatch between the patch builds have? GENCODE v31 is built against <code>GRCh38.p12</code> (see the in the title on the <a href="https://www.gencodegenes.org/human/release_31.html">webpage</a>), but obviously the no alt analysis set is derived from <code>GRCh38.p0</code> (with the pseudoautosomal regions hard masked, the EBV chromosome added, etc.)</p>
</blockquote>
<p>As is apparent in the question, the <code>GRCh38</code> reference genome is regularly patched with non-coordinate altering, backwards compatable changes (see more information <a href="https://www.ncbi.nlm.nih.gov/grc/help/patches/">here</a>). On the other hand, each <code>GENCODE</code> gene model release is based on a particular patch of the reference genome. This may be problematic because most bioinformatics analyses use a reference genome based on the non-patched version of <code>GRCh38</code>. What follows is our discussion and investigation into the effect between mismatching nucleotide sequences in the reference genome and the gene model.</p>
<p>First, we researched what some of the projects we respect in the community are doing:</p>
<div class="table-wrapper"><table><thead><tr><th>Pipeline</th><th>Reference Genome</th><th>Reference Genome Patch</th><th>Gene Model</th><th>Gene Model Patch</th></tr></thead><tbody>
<tr><td>GDC's <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/">mRNA-Seq pipeline</a></td><td><a href="https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files"><code>GRCh38_no_alt</code>-based w/ decoys + viral</a></td><td><code>GRCh38.p0</code></td><td><a href="https://www.gencodegenes.org/human/release_22.html">GENCODE v22</a></td><td><code>GRCh38.p2</code></td></tr>
<tr><td>ENCODE's <a href="https://www.encodeproject.org/pipelines/ENCPL002LPE/https://www.encodeproject.org/pages/pipelines/#RNA-seq">RNA-Seq pipeline</a></td><td><a href="https://www.encodeproject.org/files/ENCFF742NER/"><code>GRCh38_no_alt</code>-based w/ SpikeIns</a></td><td><code>GRCh38.p0</code></td><td><a href="https://www.gencodegenes.org/human/release_24.html">GENCODE v24</a></td><td><code>GRCh38.p5</code></td></tr>
<tr><td>Broad Institute's <a href="https://github.com/broadinstitute/gtex-pipeline/tree/master/rnaseq#reference-genome-and-annotation">GTEx + TOPMed RNA-Seq pipeline</a></td><td><a href="https://software.broadinstitute.org/gatk/download/bundle">Broad's <code>GRCh38</code> w/ ERCC SpikeIn</a></td><td><code>GRCh38.p0</code></td><td><a href="https://www.gencodegenes.org/human/release_26.html">GENCODE v26</a></td><td><code>GRCh38.p10</code></td></tr>
</tbody></table>
</div>
<p><strong>Note:</strong> You can confirm which patch the GENCODE genesets is based on just by clicking on the hyperlink. Verifying that each of these reference genomes is really based on <code>GRCh38_no_alt</code> takes a little bit more elbow grease: if you're interested, you can check out the comparison table <a href="0001-rnaseq-workflow-v2.0.0.html#reference-genome-comparison">in the appendix</a>. If you are <em>really</em> interested, you can recapitulate those results by running <a href="https://github.com/stjudecloud/rfcs/tree/master/resources/0001-rnaseq-workflow-v2.0.0/GenomeComparison.ipynb">the associated Jupyter notebook</a>.</p>
<p>Based on the results of the above investigation, I reached out to the author of STAR, Alex Dobin, to get his opinion on whether the differences might affect some results. You can read my question and his reply <a href="https://github.com/alexdobin/STAR/issues/673">here</a>. In short, he confirms that, yes, this may alter results for the various analysis types we were interested in.</p>
<p>Given the landscape of the community and the author's response, we considered three possible options for moving ahead:</p>
<ol>
<li>We could try using the reference FASTA supplied with the respective GENCODE release as suggested by Dr. Dobin. This was the most undesirable approach from our groups perspective:
<ul>
<li>The main reason was that this reference genome did not mirror the sequence dictionary or characteristics of the reference genome we use in our DNA-Seq pipeline out of the box. As outlined in <a href="0001-rnaseq-workflow-v2.0.0.html#Motivation">the motivation section</a> of this document, this incongruency was a major driving factor for the creation of this RFC.</li>
<li>We agreed it would require too large an amount of postprocessing of the GENCODE reference genome to convert to apply all of the no alt analysis set changes (e.g. converting to UCSC names, masking regions of the genome, inserting the EBV chromosome).</li>
<li>Additionally, this could leave room for the introduction of lots of strange errors, and there was no interest in getting into the business of genome generation (there is a reason no one applies the patches to their no alt analysis set).</li>
</ul>
</li>
<li>We could downgrade the GENCODE gene model to the latest release that was still based on <code>GRCh38.p0</code> (<a href="https://www.gencodegenes.org/human/release_21.html">GENCODE v21</a> would the correct version to use in this case).
<ul>
<li>The concordance of the reference sequences obviously made this choice an attractive option. However, <code>GENCODE v21</code> was released over 5 years ago (06.2014) and there were many valuable updates over that time. In particular, a quick look showed that there were many more transcripts and that the number of lncRNAs more than doubled (see appendix). We did not want to lose all of this forward progress if we could avoid it. To see what we looked at, you can see <a href="0001-rnaseq-workflow-v2.0.0.html#GENCODE-feature-comparisons">the relevant section in the appendix</a>.</li>
</ul>
</li>
<li>We could use the latest GENCODE release despite the mismatches outlined above as it appears most other projects have done.
<ul>
<li>The general consensus was that the effect of this discordance would be small enough to tolerate so that we could gain all of the knowledge accumulated since the older release.</li>
<li>To quantify this, we measured the differences in gene expression (as measured by the <code>R^2</code> value between <code>GENCODE v21</code> and <code>GENCODE v31</code>) and splice junction detection (as measured by the number of splice junctions detected and the relative proportions of novel/partially novel/known junctions).</li>
</ul>
</li>
</ol>
<p>After discussion with the group, we decided to stick with option #3.</p>
<h2 id="gene-model-post-processing"><a class="header" href="#gene-model-post-processing">Gene model post-processing</a></h2>
<p>Originally, I had posed this question to the group:</p>
<blockquote>
<ul>
<li>Previously, we were filtering out anything not matching &quot;level 1&quot; or &quot;level 2&quot; from the gene model. This was due to best practices outlined during our RNA-Seq Workflow v1.0 discussions. I propose we revert this for the following reasons:
<ul>
<li>The first sentence in section 2.2.2 of the <a href="https://github.com/alexdobin/STAR/blob/2.7.1a/doc/STARmanual.pdf">STAR 2.7.1.a manual</a>: &quot;The use of the most comprehensive annotations for a given species is strongly recommended&quot;. So it seems the author recommends you use the most comprehensive gene model.</li>
<li>Here is what <a href="https://www.gencodegenes.org/pages/faq.html">the GENCODE FAQ</a> has to say about the level 3 annotations: &quot;Ensembl loci where they are different from the Havana annotation or where no Havana annotation can be found&quot;. Given that the GENCODE geneset is the union of automated annotations from the <code>Ensembl-genebuild</code> and manual curation of the <code>Ensembl-Havana</code> team, this level should be harmless in the event that levels 1 &amp; 2 don't apply.</li>
<li>Last, the various other pipelines in the community don't tend to remove these features:
<ul>
<li>The <a href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#rna-seq-alignment-command-line-parameters">GDC documentation</a> does not currently describe any filtering of their <a href="https://www.gencodegenes.org/human/release_22.html">GENCODE v22</a> GTF (see the section labeled &quot;Step 1: Building the STAR index&quot;).</li>
<li>ENCODE is not filtering any features, they are just changing some of the contig names in the <a href="https://www.gencodegenes.org/human/release_24.html">GENCODE v24</a> GTF (see the analysis <a href="0001-rnaseq-workflow-v2.0.0.html#ENCODE-GTF-generation">in the appendix</a>).</li>
<li>The TOPMed pipeline <a href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md#reference-annotation">does outline some postprocessing</a> they are doing to the <a href="https://www.gencodegenes.org/human/release_26.html">GENCODE v26</a> GTF, but it's mostly just collapsing exons. I inspected the script, which does have some functionality built in to blacklist a list of transcript IDs. However, the documentation does not specify that this is turned on by default.</li>
</ul>
</li>
</ul>
</li>
</ul>
</blockquote>
<p>After discussion internally, we decided to discontinue removing <code>level 3</code> annotations by default. This is more consistent with what is being done in the community and it was decided that this was the most straightforward method with little associated risk. Therefore we are no longer performing any post-processing of the gene model.</p>
<h2 id="quality-of-life-improvements"><a class="header" href="#quality-of-life-improvements">Quality of life improvements</a></h2>
<ul>
<li>Add <code>picard ValidateSamFile</code> to the checks after the <code>STAR</code> alignment and <code>picard SortSam</code> steps. The criticism internally is that <code>ValidateSamFile</code> is quite stringent and often errors with concerns we don't care about. I'm testing this out as I develop the pipeline, and so far, I've found the following warnings to be ignore-worthy:
<ul>
<li><code>INVALID_PLATFORM_VALUE</code> is pretty annoying. It just complains if a read group doesn't contain a <code>PL</code> attribute. I'm not sure it's worth going back and fixing these.</li>
<li><code>MISSING_PLATFORM_VALUE</code>. Similar to <code>INVALID_PLATFORM_VALUE</code>, some of our samples have read groups with missing platform values and so we ignore those errors for now.</li>
</ul>
</li>
<li>For dependency management, we have moved to using <code>conda</code> within standard docker images. All packages should be available within the <code>defaults</code>, <code>conda-forge</code>, and <code>bioconda</code> repositories.</li>
<li>Add a checksum algorithm and publish the results in the data browser. After consideration internally, we decided to use the <code>md5sum</code> checksum because of its ubiquity.</li>
</ul>
<h2 id="various-other-changes"><a class="header" href="#various-other-changes">Various other changes</a></h2>
<ul>
<li>Removed a section of the pipeline that reformatted the header of the BAM to be cleaner. <code>STAR</code> outputs a header that is formatted fine already, and I found this code to just be an area where an error could be introduced for little benefit.</li>
<li>Removed a section of custom code that checks for duplicate read groups. <code>picard ValidateSamFile</code> does this for you (see <a href="https://software.broadinstitute.org/gatk/documentation/article.php?id=7571">the documentation</a> for this tool. Specifically, the <code>DUPLICATE_READ_GROUP_ID</code> error).</li>
</ul>
<h1 id="specification"><a class="header" href="#specification">Specification</a></h1>
<h2 id="dependencies"><a class="header" href="#dependencies">Dependencies</a></h2>
<p>If you'd like the full <code>conda</code> environment, you can install it using the following command. Obviously, you'll need to install <a href="https://www.anaconda.com/">anaconda</a> first.</p>
<pre><code class="language-bash">conda create -n star-mapping \
    -c conda-forge \
    -c bioconda \
    picard==2.20.2 \
    samtools==1.9 \
    star==2.7.1a \
    htseq==0.11.2 \
    ngsderive==1.0.2 \
    -y
</code></pre>
<p>Additionally, you will want to install our <code>fqlib</code> library to check that FastQ files are properly paired and have no duplicate read names. Installation of the <a href="https://rustup.rs/">Rust</a> programming language is required.</p>
<pre><code class="language-bash">cargo install --git https://github.com/stjude/fqlib.git --tag v0.3.1
</code></pre>
<h2 id="reference-files"><a class="header" href="#reference-files">Reference files</a></h2>
<p>The following reference files are used as the basis of the RNA-Seq Workflow v2.0.0:</p>
<ul>
<li>
<p>Similarly to all analysis pipelines in St. Jude Cloud, we use the <code>GRCh38_no_alt</code> analysis set for our reference genome. You can get a copy of the file <a href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz">here</a>. Additionally, you can get the file by running the following commands:</p>
<pre><code class="language-bash">wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz -O GRCh38_no_alt.fa.gz
gunzip GRCh38_no_alt.fa.gz

echo &quot;a6da8681616c05eb542f1d91606a7b2f  GRCh38_no_alt.fa&quot; &gt; GRCh38_no_alt.fa.md5
md5sum -c GRCh38_no_alt.fa.md5
# &gt; GRCh38_no_alt.fa: OK
</code></pre>
</li>
<li>
<p>For the gene model, we use the GENCODE v31 &quot;comprehensive gene annotation&quot; GTF for the &quot;CHR&quot; regions. You can get a copy of the gene annotation file <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gtf.gz">here</a>. For the exact steps to generate the gene model we use, you can run the following commands:</p>
<pre><code class="language-bash">wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.annotation.gtf.gz
echo &quot;0aac86e8a6c9e5fa5afe2a286325da81  gencode.v31.annotation.gtf.gz&quot; &gt; gencode.v31.annotation.gtf.gz.md5
md5sum -c gencode.v31.annotation.gtf.gz.md5
# &gt; gencode.v31.annotation.gtf.gz: OK

gunzip gencode.v31.annotation.gtf.gz
echo &quot;4e22351ae216e72aa57cd6d6011960f8  gencode.v31.annotation.gtf&quot; &gt; gencode.v31.annotation.gtf.md5
md5sum -c gencode.v31.annotation.gtf.md5
# &gt; gencode.v31.annotation.gtf: OK
</code></pre>
</li>
<li>
<p>Last, the following command is used to prepare the STAR index file:</p>
<pre><code class="language-bash">STAR --runMode genomeGenerate \                    # Use genome generation runMode.
     --genomeDir $OUTPUT_DIR \                     # Specify an output directory.
     --runThreadN $NCPU \                          # Number of threads to use to build genome database.
     --genomeFastaFiles $FASTA \                   # A path to the GRCh38_no_alt.fa FASTA file.
     --sjdbGTFfile $GENCODE_GTF_V31 \              # GENCODE v31 gene model file.
     --sjdbOverhang 125                            # Splice junction database overhang parameter, the optimal value is (Max length of RNA-Seq read-1).
</code></pre>
</li>
</ul>
<h2 id="workflow"><a class="header" href="#workflow">Workflow</a></h2>
<p>Here are the resulting steps in the RNA-Seq Workflow v2.0.0 pipeline. There might be slight alterations in the actual implementation, which can be found in <a href="https://github.com/stjudecloud/workflows/blob/master/workflows/rnaseq/rnaseq-standard.wdl">the St. Jude Cloud workflows repository</a>.</p>
<ol>
<li>
<p>Run <code>picard ValidateSam</code> on the incoming BAM to ensure that it is well-formed enough to convert back to FastQ.</p>
<pre><code class="language-bash">picard ValidateSamFile I=$INPUT_BAM \                # Input BAM.
                  IGNORE=INVALID_PLATFORM_VALUE \    # Validations to ignore.
                  IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Split BAM file into multiple BAMs on the different read groups using <code>samtools split</code>. See <a href="http://www.htslib.org/doc/samtools.html">the samtools documentation</a> for more information.</p>
<pre><code class="language-bash">samtools split -u $UNACCOUNTED_BAM_NAME \ # Reads that do not belong to a read group or the read group is unrecognized go here.
               -f '%*_%!.%.'              # Format of output BAM file names.
</code></pre>
<p>If the BAM has unaccounted reads, those will need to be triaged and this step will need to be rerun.</p>
</li>
<li>
<p>Run Picard <code>SamToFastq</code> on each of the BAMs generated in the previous step.</p>
<pre><code class="language-bash">   picard SamToFastq \
          INPUT=$INPUT_BAM \
          FASTQ=$FASTQ_R1 \
          SECOND_END_FASTQ=$FASTQ_R2 \
          RE_REVERSE=true \
          VALIDATION_STRINGENCY=SILENT
</code></pre>
</li>
<li>
<p>Run <code>fq lint</code> on each of the FastQ pairs that was generated in the previous step as a sanity check. You can see the checks that the <code>fq</code> tool performs <a href="https://github.com/stjude/fqlib/blob/master/README.md#validators">here</a>.</p>
<pre><code class="language-bash">fq lint $FASTQ_R1 $FASTQ_R2 # Files for read 1 and read 2.
</code></pre>
</li>
<li>
<p>Run the <code>STAR</code> alignment algorithm.</p>
<pre><code class="language-bash">STAR --readFilesIn $ALL_FASTQ_R1 $ALL_FASTQ_READ2 \ # FastQ files, separated by comma if there are multiple. The order of your R1 and R2 files has to match!
     --outSAMattrRGline $ALL_RG \                   # Read group lines in the same order as `readFilesIn` (derived from earlier `samtools split` step).
     --genomeDir $STARDB \                          # Directory containing the STAR genome
     --runThreadN $NCPU \                           # Number of threads to use. You must request the correct amount from HPCF first!
     --outSAMunmapped Within \                      # Keep unmapped reads in the final BAM file.
     --outSAMstrandField intronMotif \              # Preserve compatibility with Cufflinks by including the XS attribute (strand derived from intron motif).
     --outSAMattributes NH HI AS nM NM MD XS \      # Recommended SAM attributes to include for compatibility. Refer to manual for specifics.
     --outFilterMultimapScoreRange 1 \              # Ensures that all multi-mapped reads will need to share the mapping score.
     --outFilterMultimapNmax 20 \                   # Max number of multi-mapped SAM entries for each read.
     --outFilterMismatchNmax 10 \                   # Max number of mismatches allowed from alignment.
     --alignIntronMax 500000 \                      # Max intron size considered.
     --alignMatesGapMax 1000000 \                   # Max gap allowed between two mates.
     --sjdbScore 2 \                                # Additional weight given to alignments that cross database junctions when scoring alignments.
     --alignSJDBoverhangMin 1 \                     # Minimum overhang required on each side of a splice junction. Here, we require only one base on each side.
     --outFilterMatchNminOverLread 0.66 \           # 66% of the read must be perfectly matched to the reference sequence.
     --outFilterScoreMinOverLread 0.66 \            # Score must be greater than 66% of the read length. So for RL=100, the alignment must have a score &gt; 66.
     --limitBAMsortRAM $RAM_LIMIT \                 # Amount of RAM to use for sorting. Recommended value is [Max amount of RAM] - 5GB.
     --outFileNamePrefix $OUT_FILE_PREFIX \         # All output files will have this path prepended.
     --twopassMode Basic                            # Use STAR two-pass mapping technique (refer to manual).
</code></pre>
</li>
<li>
<p>Run <code>picard SortSam</code> on the <code>STAR</code>-aligned BAM file. Note that this is much more memory efficient than using <code>STAR</code>'s built-in sorting (which often takes 100GB+ of RAM).</p>
<pre><code class="language-bash">picard SortSam I=$STAR_BAM \                  # Input BAM.
               O=$SORTED_BAM \                # Coordinate-sorted BAM.
               SO=&quot;coordinate&quot; \              # Specify the output should be coordinate-sorted
               CREATE_INDEX=false \           # Explicitly do not create an index at this step, in case the default changes.
               CREATE_MD5_FILE=false \        # Explicity do not create an md5 checksum at this step, in case the default changes.
               COMPRESSION_LEVEL=5 \          # Explicitly set the compression level to 5, although, at the time of writing, this is the default.
               VALIDATION_STRINGENCY=SILENT   # Turn of validation stringency for this step.
</code></pre>
</li>
<li>
<p>Index the coordinate-sorted BAM file.</p>
<pre><code class="language-bash">samtools index $STAR_SORTED_BAM # STAR-aligned, coordinate-sorted BAM.
</code></pre>
</li>
<li>
<p>Run <code>picard ValidateSamFile</code> on the aligned and sorted BAM file.</p>
<pre><code class="language-bash">picard ValidateSamFile I=$STAR_SORTED_BAM \    # STAR-aligned, coordinate-sorted BAM.
               IGNORE=INVALID_PLATFORM_VALUE \ # Validations to ignore.
               IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Run <code>ngsderive</code>'s <code>strandedness</code> subcommand to confirm that the lab's information on strandedness reflects what was is computed. Manually triage any discrepancies. This is particularly useful for historical samples. Additionally,
if the value for strandedness isn't known at run time, we can use the inferred value (if reported).</p>
<pre><code class="language-bash">ngsderive strandedness $STAR_SORTED_BAM \     # STAR-aligned, coordinate-sorted BAM.
                       -g $GENCODE_GTF_V31_GZ # GENCODE v31 GTF (gzipped)
</code></pre>
</li>
<li>
<p>Next, <code>htseq-count</code> is run for the final counts file to be delivered.</p>
<pre><code class="language-bash">htseq-count -f bam \                            # Specify input file as BAM.
           -r pos \                             # Specify the BAM is position sorted.
           -s $PROVIDED_OR_INFERRED \           # Strandedness as specified by the lab and confirmed by `ngsderive strandedness` above. Typically `reverse` for St. Jude Cloud data.
           -m union \                           # For reference, GDC uses &quot;intersection-nonempty&quot;.
           -i gene_name \                       # I'd like to use the colloquial gene name here. For reference, GDC uses &quot;gene_id&quot; here. Needs input from reviewers.
           --secondary-alignments ignore \      # Elect to ignore secondary alignments. Needs input from reviewers.
           --supplementary-alignments ignore \  # Elect to ignore supplementary alignments. Needs input from reviewers.
           $STAR_SORTED_BAM \                   # STAR-aligned, coordinate-sorted BAM.
           $GENCODE_GTF_V31                     # GENCODE v31 GTF
</code></pre>
</li>
</ol>
<h1 id="appendix"><a class="header" href="#appendix">Appendix</a></h1>
<h3 id="reference-genome-comparison"><a class="header" href="#reference-genome-comparison">Reference genome comparison</a></h3>
<p>Below are the results of an analysis of each pipeline's <code>GRCh38</code>-based reference genome. The steps are essentially:</p>
<ol>
<li>Research what reference genome each project is using and where to find it.</li>
<li>Download it.</li>
<li>Use <code>picard CreateSequenceDictionary</code> to get the md5sums for each sequence in the dictionary.</li>
<li>Compare for the common chromosomes in each reference (the autosomes, the sex chromosomes, and the EBV decoy).</li>
</ol>
<p>If you are interested in seeing the <em>full</em> comparison table or in regenerating these results, you can see <a href="https://github.com/stjudecloud/rfcs/tree/master/resources/0001-rnaseq-workflow-v2.0.0/GenomeComparison.ipynb">the associated Jupyter notebook</a>.</p>
<div class="table-wrapper"><table><thead><tr><th>Sequence Name</th><th>NCBI (baseline)</th><th>ENCODE</th><th>GDC</th><th>TOPMed</th><th>Concordant</th></tr></thead><tbody>
<tr><td>chr1</td><td><code>6aef897c3d6ff0c78aff06ac189178dd</code></td><td><code>6aef897c3d6ff0c78aff06ac189178dd</code></td><td><code>6aef897c3d6ff0c78aff06ac189178dd</code></td><td><code>6aef897c3d6ff0c78aff06ac189178dd</code></td><td>True</td></tr>
<tr><td>chr2</td><td><code>f98db672eb0993dcfdabafe2a882905c</code></td><td><code>f98db672eb0993dcfdabafe2a882905c</code></td><td><code>f98db672eb0993dcfdabafe2a882905c</code></td><td><code>f98db672eb0993dcfdabafe2a882905c</code></td><td>True</td></tr>
<tr><td>chr3</td><td><code>76635a41ea913a405ded820447d067b0</code></td><td><code>76635a41ea913a405ded820447d067b0</code></td><td><code>76635a41ea913a405ded820447d067b0</code></td><td><code>76635a41ea913a405ded820447d067b0</code></td><td>True</td></tr>
<tr><td>chr4</td><td><code>3210fecf1eb92d5489da4346b3fddc6e</code></td><td><code>3210fecf1eb92d5489da4346b3fddc6e</code></td><td><code>3210fecf1eb92d5489da4346b3fddc6e</code></td><td><code>3210fecf1eb92d5489da4346b3fddc6e</code></td><td>True</td></tr>
<tr><td>chr5</td><td><code>a811b3dc9fe66af729dc0dddf7fa4f13</code></td><td><code>a811b3dc9fe66af729dc0dddf7fa4f13</code></td><td><code>a811b3dc9fe66af729dc0dddf7fa4f13</code></td><td><code>a811b3dc9fe66af729dc0dddf7fa4f13</code></td><td>True</td></tr>
<tr><td>chr6</td><td><code>5691468a67c7e7a7b5f2a3a683792c29</code></td><td><code>5691468a67c7e7a7b5f2a3a683792c29</code></td><td><code>5691468a67c7e7a7b5f2a3a683792c29</code></td><td><code>5691468a67c7e7a7b5f2a3a683792c29</code></td><td>True</td></tr>
<tr><td>chr7</td><td><code>cc044cc2256a1141212660fb07b6171e</code></td><td><code>cc044cc2256a1141212660fb07b6171e</code></td><td><code>cc044cc2256a1141212660fb07b6171e</code></td><td><code>cc044cc2256a1141212660fb07b6171e</code></td><td>True</td></tr>
<tr><td>chr8</td><td><code>c67955b5f7815a9a1edfaa15893d3616</code></td><td><code>c67955b5f7815a9a1edfaa15893d3616</code></td><td><code>c67955b5f7815a9a1edfaa15893d3616</code></td><td><code>c67955b5f7815a9a1edfaa15893d3616</code></td><td>True</td></tr>
<tr><td>chr9</td><td><code>6c198acf68b5af7b9d676dfdd531b5de</code></td><td><code>6c198acf68b5af7b9d676dfdd531b5de</code></td><td><code>6c198acf68b5af7b9d676dfdd531b5de</code></td><td><code>6c198acf68b5af7b9d676dfdd531b5de</code></td><td>True</td></tr>
<tr><td>chr10</td><td><code>c0eeee7acfdaf31b770a509bdaa6e51a</code></td><td><code>c0eeee7acfdaf31b770a509bdaa6e51a</code></td><td><code>c0eeee7acfdaf31b770a509bdaa6e51a</code></td><td><code>c0eeee7acfdaf31b770a509bdaa6e51a</code></td><td>True</td></tr>
<tr><td>chr11</td><td><code>1511375dc2dd1b633af8cf439ae90cec</code></td><td><code>1511375dc2dd1b633af8cf439ae90cec</code></td><td><code>1511375dc2dd1b633af8cf439ae90cec</code></td><td><code>1511375dc2dd1b633af8cf439ae90cec</code></td><td>True</td></tr>
<tr><td>chr12</td><td><code>96e414eace405d8c27a6d35ba19df56f</code></td><td><code>96e414eace405d8c27a6d35ba19df56f</code></td><td><code>96e414eace405d8c27a6d35ba19df56f</code></td><td><code>96e414eace405d8c27a6d35ba19df56f</code></td><td>True</td></tr>
<tr><td>chr13</td><td><code>a5437debe2ef9c9ef8f3ea2874ae1d82</code></td><td><code>a5437debe2ef9c9ef8f3ea2874ae1d82</code></td><td><code>a5437debe2ef9c9ef8f3ea2874ae1d82</code></td><td><code>a5437debe2ef9c9ef8f3ea2874ae1d82</code></td><td>True</td></tr>
<tr><td>chr14</td><td><code>e0f0eecc3bcab6178c62b6211565c807</code></td><td><code>e0f0eecc3bcab6178c62b6211565c807</code></td><td><code>e0f0eecc3bcab6178c62b6211565c807</code></td><td><code>e0f0eecc3bcab6178c62b6211565c807</code></td><td>True</td></tr>
<tr><td>chr15</td><td><code>f036bd11158407596ca6bf3581454706</code></td><td><code>f036bd11158407596ca6bf3581454706</code></td><td><code>f036bd11158407596ca6bf3581454706</code></td><td><code>f036bd11158407596ca6bf3581454706</code></td><td>True</td></tr>
<tr><td>chr16</td><td><code>db2d37c8b7d019caaf2dd64ba3a6f33a</code></td><td><code>db2d37c8b7d019caaf2dd64ba3a6f33a</code></td><td><code>db2d37c8b7d019caaf2dd64ba3a6f33a</code></td><td><code>db2d37c8b7d019caaf2dd64ba3a6f33a</code></td><td>True</td></tr>
<tr><td>chr17</td><td><code>f9a0fb01553adb183568e3eb9d8626db</code></td><td><code>f9a0fb01553adb183568e3eb9d8626db</code></td><td><code>f9a0fb01553adb183568e3eb9d8626db</code></td><td><code>f9a0fb01553adb183568e3eb9d8626db</code></td><td>True</td></tr>
<tr><td>chr18</td><td><code>11eeaa801f6b0e2e36a1138616b8ee9a</code></td><td><code>11eeaa801f6b0e2e36a1138616b8ee9a</code></td><td><code>11eeaa801f6b0e2e36a1138616b8ee9a</code></td><td><code>11eeaa801f6b0e2e36a1138616b8ee9a</code></td><td>True</td></tr>
<tr><td>chr19</td><td><code>85f9f4fc152c58cb7913c06d6b98573a</code></td><td><code>85f9f4fc152c58cb7913c06d6b98573a</code></td><td><code>85f9f4fc152c58cb7913c06d6b98573a</code></td><td><code>85f9f4fc152c58cb7913c06d6b98573a</code></td><td>True</td></tr>
<tr><td>chr20</td><td><code>b18e6c531b0bd70e949a7fc20859cb01</code></td><td><code>b18e6c531b0bd70e949a7fc20859cb01</code></td><td><code>b18e6c531b0bd70e949a7fc20859cb01</code></td><td><code>b18e6c531b0bd70e949a7fc20859cb01</code></td><td>True</td></tr>
<tr><td>chr21</td><td><code>974dc7aec0b755b19f031418fdedf293</code></td><td><code>974dc7aec0b755b19f031418fdedf293</code></td><td><code>974dc7aec0b755b19f031418fdedf293</code></td><td><code>974dc7aec0b755b19f031418fdedf293</code></td><td>True</td></tr>
<tr><td>chr22</td><td><code>ac37ec46683600f808cdd41eac1d55cd</code></td><td><code>ac37ec46683600f808cdd41eac1d55cd</code></td><td><code>ac37ec46683600f808cdd41eac1d55cd</code></td><td><code>ac37ec46683600f808cdd41eac1d55cd</code></td><td>True</td></tr>
<tr><td>chrX</td><td><code>2b3a55ff7f58eb308420c8a9b11cac50</code></td><td><code>2b3a55ff7f58eb308420c8a9b11cac50</code></td><td><code>2b3a55ff7f58eb308420c8a9b11cac50</code></td><td><code>2b3a55ff7f58eb308420c8a9b11cac50</code></td><td>True</td></tr>
<tr><td>chrY</td><td><code>ce3e31103314a704255f3cd90369ecce</code></td><td><code>ce3e31103314a704255f3cd90369ecce</code></td><td><code>ce3e31103314a704255f3cd90369ecce</code></td><td><code>ce3e31103314a704255f3cd90369ecce</code></td><td>True</td></tr>
<tr><td>chrM</td><td><code>c68f52674c9fb33aef52dcf399755519</code></td><td><code>c68f52674c9fb33aef52dcf399755519</code></td><td><code>c68f52674c9fb33aef52dcf399755519</code></td><td><code>c68f52674c9fb33aef52dcf399755519</code></td><td>True</td></tr>
<tr><td>chrEBV</td><td><code>6743bd63b3ff2b5b8985d8933c53290a</code></td><td><code>6743bd63b3ff2b5b8985d8933c53290a</code></td><td><code>6743bd63b3ff2b5b8985d8933c53290a</code></td><td><code>6743bd63b3ff2b5b8985d8933c53290a</code></td><td>True</td></tr>
</tbody></table>
</div>
<h3 id="encode-gtf-generation"><a class="header" href="#encode-gtf-generation">ENCODE GTF generation</a></h3>
<p>A small investigation was done to see how ENCODE's GENCODE v24 GTF was being produced (the md5sum between the original GENCODE v24 primary assembly file and ENCODE's version of the file weren't matching up). In short, they just changed the names of some of the contigs. You can verify this yourself by doing an <code>md5sum</code> without the sequence name column of the GTF (or manually download and inspect the differences like I did :)).</p>
<pre><code class="language-bash">curl -sL ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_24/gencode.v24.primary_assembly.annotation.gtf.gz | \
         gunzip -c | \
         cut -f2- -d$'\t' | \
         md5sum
# 5227fac91c8d32b3fa3a8f78c4bf0e5c  -

curl -sL https://www.encodeproject.org/files/gencode.v24.primary_assembly.annotation/@@download/gencode.v24.primary_assembly.annotation.gtf.gz | \
         gunzip -c | \
         cut -f2- -d$'\t' | \
         md5sum
# 5227fac91c8d32b3fa3a8f78c4bf0e5c  -
</code></pre>
<h3 id="gencode-feature-comparisons"><a class="header" href="#gencode-feature-comparisons">GENCODE feature comparisons</a></h3>
<p>Below are a few commands used to quickly evaluate how much the GENCODE geneset has changed over time. This was useful in our discussion about how much value would be lost if we just used <code>GENCODE v21</code> (which was based on <code>GRCh38.p0</code>). See <a href="0001-rnaseq-workflow-v2.0.0.html#GENCODE-compatability">the discussion on the GENCODE compatibility</a> for more information.</p>
<ul>
<li>
<p>Commands that were used in the comparison of feature types between <code>GENCODE v21</code> and <code>GENCODE v31</code>:</p>
<pre><code class="language-bash">gawk '$0 !~ /^#/ { features[$3] += 1; total += 1 } END {for (key in features) { print &quot;Feature &quot; key &quot;: &quot; features[key] &quot; (&quot; (features[key]/total)*100 &quot;%)&quot; }; print &quot;Total: &quot; total}' gencode.v21.annotation.gtf
# Feature exon: 1162114 (45.6341%)
# Feature CDS: 696929 (27.3671%)
# Feature UTR: 275100 (10.8027%)
# Feature gene: 60155 (2.36217%)
# Feature start_codon: 81862 (3.21457%)
# Feature Selenocysteine: 114 (0.00447657%)
# Feature stop_codon: 73993 (2.90557%)
# Feature transcript: 196327 (7.7094%)
# Total: 2546594

gawk '$0 !~ /^#/ { features[$3] += 1; total += 1 } END {for (key in features) { print &quot;Feature &quot; key &quot;: &quot; features[key] &quot; (&quot; (features[key]/total)*100 &quot;%)&quot; }; print &quot;Total: &quot; total}' gencode.v31.annotation.gtf
# Feature exon: 1363843 (47.3247%)
# Feature CDS: 755320 (26.2092%)
# Feature UTR: 308315 (10.6984%)
# Feature gene: 60603 (2.10289%)
# Feature start_codon: 87299 (3.02923%)
# Feature Selenocysteine: 119 (0.00412924%)
# Feature stop_codon: 79505 (2.75878%)
# Feature transcript: 226882 (7.87269%)
# Total: 2881886
</code></pre>
</li>
<li>
<p>Commands that were used in the comparison of <code>gene_types</code> for <code>gene</code> features between <code>GENCODE v21</code> and <code>GENCODE v31</code>:</p>
<pre><code class="language-bash">gawk '$3 ~ /gene/' gencode.v21.annotation.gtf | gawk 'match($0, /gene_type &quot;([A-Za-z0-9]+)&quot;/, a) { types[a[1]] += 1; total += 1 } END {for (key in types) { print &quot;Gene type &quot; key &quot;: &quot; types[key] &quot; (&quot; (types[key]/total)*100 &quot;%)&quot; }; print &quot;Total: &quot; total}'
# Gene type rRNA: 549 (2.54508%)
# Gene type antisense: 5542 (25.6919%)
# Gene type pseudogene: 29 (0.13444%)
# Gene type lincRNA: 7666 (35.5385%)
# Gene type TEC: 1058 (4.90473%)
# Gene type miRNA: 3837 (17.7878%)
# Gene type snoRNA: 978 (4.53386%)
# Gene type snRNA: 1912 (8.86375%)
# Total: 21571

gawk '$3 ~ /gene/' gencode.v31.annotation.gtf | gawk 'match($0, /gene_type &quot;([A-Za-z0-9]+)&quot;/, a) { types[a[1]] += 1; total += 1 } END {for (key in types) { print &quot;Gene type &quot; key &quot;: &quot; types[key] &quot; (&quot; (types[key]/total)*100 &quot;%)&quot; }; print &quot;Total: &quot; total}'
# Gene type ribozyme: 8 (0.0351463%)
# Gene type scaRNA: 49 (0.215271%)
# Gene type lncRNA: 16840 (73.983%)
# Gene type rRNA: 52 (0.228451%)
# Gene type vaultRNA: 1 (0.00439329%)
# Gene type scRNA: 1 (0.00439329%)
# Gene type sRNA: 5 (0.0219664%)
# Gene type pseudogene: 18 (0.0790792%)
# Gene type TEC: 1064 (4.67446%)
# Gene type miRNA: 1881 (8.26377%)
# Gene type snoRNA: 942 (4.13848%)
# Gene type snRNA: 1901 (8.35164%)
# Total: 22762
</code></pre>
</li>
<li>
<p>The following is a quick command that can be used on a GENCODE GTF to summarize the level counts/percentages in the GTF file. This was used to quantify how much of the GTF would be eliminated when all level 3 features were removed.</p>
<pre><code class="language-bash">gawk 'match($0, /level ([0-9]+)/, a) { levels[a[1]] += 1; total += 1 } END {for (key in levels) { print &quot;Level &quot; key &quot;: &quot; levels[key] &quot; (&quot; (levels[key]/total)*100 &quot;%)&quot; }; print &quot;Total  : &quot; total}' gencode.v31.annotation.gtf
# Level 1: 186461 (6.4701%)
# Level 2: 2450972 (85.0475%)
# Level 3: 244453 (8.4824%)
# Total  : 2881886
</code></pre>
</li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="table-of-contents-1"><a class="header" href="#table-of-contents-1">Table of Contents <!-- omit in toc --></a></h1>
<ul>
<li><a href="0002-quality-check-workflow.html#introduction">Introduction</a></li>
<li><a href="0002-quality-check-workflow.html#motivation">Motivation</a></li>
<li><a href="0002-quality-check-workflow.html#discussion">Discussion</a></li>
<li><a href="0002-quality-check-workflow.html#specification">Specification</a></li>
</ul>
<h2 id="introduction-1"><a class="header" href="#introduction-1">Introduction</a></h2>
<p>This RFC documents an automated workflow for assessing the integrity and quality of St. Jude Cloud omics data. The goal of this RFC is two-fold.</p>
<ol>
<li>Establish state of the art method for comprehensively evaluating genomics
data quality at scale, both at time of receipt and after processing, for
whole genome, whole exome, and transcriptome sequencing.</li>
<li>Publish a collection of metrics that end-users of St. Jude Cloud can leverage
to assess the quality of the data available. This context should save users
time computing the information themselves while also informing appropriate
use of the data.</li>
</ol>
<p>You can find the relevant discussion on the <a href="https://github.com/stjudecloud/rfcs/pull/3">associated pull request</a>.</p>
<h2 id="motivation-1"><a class="header" href="#motivation-1">Motivation</a></h2>
<p>St. Jude Cloud is a large repository of omics data available for request from
the academic and non-profit community. As such, the project processes thousands
of samples from whole genome, whole exome, RNA-Seq, and various omics-based
assays each year. A standard, robust method to assess pre-processing and
post-processing quality for samples has been developed in-house, but there are
some shortcomings with our current approach. In particular, this RFC will:</p>
<ul>
<li>Define the standard set of QC tools used to evaluating omics-based data,</li>
<li>identify and implement key metrics that can be automated to assist in manual observation of the data, and</li>
<li>outline mechanisms to publish those QC results alongside the data already in
St. Jude Cloud so that end-users can leverage this information.</li>
</ul>
<h2 id="discussion-1"><a class="header" href="#discussion-1">Discussion</a></h2>
<h3 id="types-of-qc"><a class="header" href="#types-of-qc">Types of QC</a></h3>
<p>There are (at least) two different types of QC typically carried out omics-based data. <strong>Experimental</strong> QC attempts to identify the success of the assay(s) performed. Once one is sufficiently satisfied that the data generated by an experiment is &quot;good&quot;, <strong>computational</strong> QC examines the degree to which computational processing of that data was completed successfully.</p>
<p>By the time data reaches the St. Jude Cloud team from various sources, extensive
<em>experimental</em> and <em>computational</em> evaluation have already been carried out.
Each contributing project has its own thresholds for quality in both areas,
which is dependent on the best practices at that point in time and the goals of
the project. Most often, our team takes the computational data, reverts it back
to its raw form (such as FastQ files), and reprocesses the data using a
harmonization pipeline.</p>
<p>Thus, the scope of this RFC, and the QC of samples on the project in general, is
limited to the <em>computational</em> QC of the files produced for publication in St.
Jude Cloud. While we do produce results that define <em>experimental</em> results (such
as <code>fastqc</code>), these are rarely used to decide which files pass or fail our
QC‚Äîthis is in recognition of the fact that the data, while not always perfect,
is extremely valuable due to its relative scarcity. We hope that the inclusion
of these results will save end-users time and aid in decision-making about
downstream analysis approaches.</p>
<h2 id="general-statistics"><a class="header" href="#general-statistics">General Statistics</a></h2>
<p>General statistics are contained within the summary table at the top of the
MultiQC reports. These statistics give a quick overview of the cohort of
interest with respect to the defined metrics. Generally, they are summary
statistics rather than being comprehensive in nature (those statistics are
included in the sections that follow after).</p>
<h3 id="summary-table"><a class="header" href="#summary-table">Summary Table</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>M Reads Mapped (<a href="0002-quality-check-workflow.html#m-reads-mapped">link</a>)</td><td><a href="http://www.htslib.org/doc/samtools.html">samtools</a></td><td>Number of reads mapped in millions. This metric is useful</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Validation Status (<a href="0002-quality-check-workflow.html#validation-status">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>The validation status of the file as determined by <code>picard ValidateSamFile</code>.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads Aligned (<a href="0002-quality-check-workflow.html#percentage-of-reads-aligned">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Number of mapped reads divided by the total number of reads as a percentage.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Median Insert Size (<a href="0002-quality-check-workflow.html#median-insert-size">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Median size of the fragment that is inserted between the sequencing adapters (estimated in silico).</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>Percentage Duplication (<a href="0002-quality-check-workflow.html#percentage-duplication">link</a>)</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Percentage of the reads that are marked as PCR or optical duplicates.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>‚â• 30X Coverage (<a href="0002-quality-check-workflow.html#-30x-coverage">link</a>)</td><td><a href="https://github.com/brentp/mosdepth">mosdepth</a></td><td>The percentage of locations that are covered by at least 30 reads for the whole genome, the exonic regions, and the coding sequence regions specifically.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/no-red" alt="No" /></td></tr>
<tr><td>Predicted Instrument (<a href="0002-quality-check-workflow.html#predicted-instrument">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted sequencing machine that produced this data. This should match your expectations for what machine sequenced the data.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Predicted Read Length (<a href="0002-quality-check-workflow.html#predicted-read-length">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted read length for the sequencing experiment. This should match your expectations for how the library was prepared. A value of -1 indicates that a stable read length could not be predicted.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Probable Encoding (<a href="0002-quality-check-workflow.html#probable-encoding">link</a>)</td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted encoding of the <em>original</em> FASTQ file. For this, all modern data should match <code>Sanger/Illumina 1.8</code>.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads attributed to Homo sapiens (<a href="0002-quality-check-workflow.html#percentage-of-reads-attributed-to-homo-sapiens">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were assigned as originating from Homo sapiens from <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads attributed to top 5 species (<a href="0002-quality-check-workflow.html#percentage-of-reads-attributed-to-top-5-species">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were assigned to the top 5 reported species from <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
<tr><td>Percentage of Reads Unclassified (<a href="0002-quality-check-workflow.html#percentage-of-reads-unclassified">link</a>)</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>The percentage of reads that were not classified by <a href="https://github.com/DerrickWood/kraken2">kraken</a>.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="Yes" /></td></tr>
</tbody></table>
</div>
<h3 id="m-reads-mapped"><a class="header" href="#m-reads-mapped"><strong>M Reads Mapped</strong></a></h3>
<p><em>Produced by <a href="http://www.htslib.org/doc/samtools.html">samtools</a></em></p>
<p>Number of reads mapped in millions. The number of reads in a whole genome
sequencing file can vary widely based on the experiment design and the
sequencing depth of the project.</p>
<ul>
<li>Typically, WGS files contain between 200M reads up to 2 billion reads (or
higher in some cases). Any deviation outside of this range should be
investigated further (particularly if the number is lower, which suggests
that the file may be truncated).</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect a normal distribution for this metric within the same
sequencing cohort, special attention is required for any samples that do not
conform to this distribution. </li>
</ul>
<h3 id="percentage-of-reads-aligned"><a class="header" href="#percentage-of-reads-aligned"><strong>Percentage of Reads Aligned</strong></a></h3>
<p><em>Produced by <a href="https://broadinstitute.github.io/picard/">picard</a></em></p>
<p>Percentage of the number of reads that were able to be mapped to a given
reference genome. Within St. Jude Cloud, this metric is indicative of the
amount of non-human material which was sequenced during the experiment.</p>
<ul>
<li>This number is especially informative within the context of samples from the
same cohort. We expect the percentage of aligned reads to be tightly
clustered above 95% mapping rate. It is not uncommon for samples to go as
low as 80% mapping rate, and this should be considered within an acceptable
range. Occasionally you may find small proportion of samples between 51%-79%
mapping rate: those samples should be investigated, but typically, you
should still include those samples in the release. Anything below 50%
mapping rate signals an issue with contamination, and anything less than 30%
mapping may be an issue with the mapper or the data integrity (e.g.
mismatched read pairs).</li>
</ul>
<h3 id="median-insert-size"><a class="header" href="#median-insert-size"><strong>Median Insert Size</strong></a></h3>
<p><em>Produced by <a href="https://broadinstitute.github.io/picard/">picard</a></em></p>
<p>Size of the fragment inserted between the sequencing adapters. In whole genome
sequencing experiments, typically you have a target fragment length for each
sample (commonly ~2x the read length to maximize the amount of information
gathered per fragment). The median insert size has proven to be an incredibly
valuable statistic for identifying mapping problems and experimental
abnormalities.</p>
<ul>
<li>Typically, the median insert size should be between 1.5x - 2x the read
length of the experiment. This criteria may be evaluated loosely, as insert
sizes that are close to this range (anything between 1x - 5x read length) do
not necessarily indicate a major problem. Samples with an extremely low
median insert size (10-40 nucleotides) are likely the cause of mapping
artifacts‚Äîparticularly if <code>bwa mem</code> was used with a low seed size.</li>
<li>Insert size distributions tend to follow a <a href="https://en.wikipedia.org/wiki/Gamma_distribution">gamma
distribution</a> where long,
rolling peaks are interspersed with tight peaks close to the target insert
size.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect the median insert size across samples to be tightly
clustered within a cohort.</li>
</ul>
<h3 id="-30x-coverage"><a class="header" href="#-30x-coverage"><strong>‚â• 30X Coverage</strong></a></h3>
<p><em>Produced by <a href="https://github.com/brentp/mosdepth">mosdepth</a></em></p>
<p>Percentage of the genome which has at least 30 reads covering each position (on
average) for the whole genome, the exonic regions, and the coding regions
(&quot;30x coverage&quot;). 30x coverage at any location is widely considered to be the
minimum number of reads needed to generate high confidence genotype call. In
whole genome sequencing, it is desirable to have as much of the genome as
possible to be covered by 30 or more reads. Expected genome coverage is often
determined at the time of sequencing and may vary from project to project
depending on cost and experimental design.</p>
<ul>
<li>Target sequencing depth and fraction of the genome to be covered are
generally set on a project by project basis. Speak with the someone on the
sequencing team to ensure you understand what sequencing depth and fraction
of the genome covered you should be expecting.</li>
<li>At a minimum, we expect 80% of the whole genome to be covered by at least 30
reads in high-quality whole genome sequencing experiments. For higher depth
coverage projects like our Clinical Genomics project, the standard is 60x
across 80% of the genome.</li>
<li>Less coverage does not necessarily indicate a problem, but in whole genome
sequencing, anything below 65% of the genome at 30X signals that something
may be going wrong. Pay close attention to the distribution of the reads
across the genome by generating a coverage plot to see where sequencing bias
is being introduced.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect this metric to be tightly clustered across samples
within a cohort.</li>
</ul>
<h3 id="gc-content"><a class="header" href="#gc-content"><strong>GC Content</strong></a></h3>
<p><em>Produced by <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></em></p>
<p>Percentage of the nucleotides contained within reads which are Cs (Cytosine) or
Gs (Guanine). GC content is important because Gs pair with Cs in DNA with 3
hydrogen bonds whereas As pair with Ts in DNA with 2 hydrogen bonds. Ergo, DNA
is considered to be stronger when comprised of more GC bonds, and GC content is
a defining characteristic of different genomes. GC content for humans is
<a href="https://www.nature.com/articles/35057062#Sec15">estimated to be at around 41%</a>.</p>
<ul>
<li>GC content in whole genome sequencing is typically between 40%-60% depending
on sequencing bias. Any deviation outside of this range should be
investigated further.</li>
<li>This number is especially informative within the context of samples from the
same cohort. We expect GC content across samples to be tightly clustered
within a cohort.</li>
<li>If either of the above assumptions do not hold true, probable sources of
variance include library preparation abnormalities or contamination issues.</li>
</ul>
<h2 id="detailed-charts"><a class="header" href="#detailed-charts">Detailed Charts</a></h2>
<h3 id="mosdepth"><a class="header" href="#mosdepth">mosdepth</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>Cumulative coverage distribution</td><td><a href="https://github.com/brentp/mosdepth">mosdepth</a></td><td>The proportion of bases within the genome with at least <code>X</code> coverage.</td><td><img src="https://img.shields.io/badge/all-brightgreen" alt="all" /></td><td><img src="https://img.shields.io/badge/exonic/cds-orange" alt="exonic/cds" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>Coverage distribution</td><td><a href="https://github.com/brentp/mosdepth">mosdepth</a></td><td>The proportion of bases within the genome with <em>exactly</em> <code>X</code> coverage.</td><td><img src="https://img.shields.io/badge/all-brightgreen" alt="all" /></td><td><img src="https://img.shields.io/badge/exonic/cds-orange" alt="exonic/cds" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>Average coverage per contig</td><td><a href="https://github.com/brentp/mosdepth">mosdepth</a></td><td>Average coverage per contig or chromosome.</td><td><img src="https://img.shields.io/badge/all-brightgreen" alt="all" /></td><td><img src="https://img.shields.io/badge/all-brightgreen" alt="all" /></td><td><img src="https://img.shields.io/badge/all-brightgreen" alt="all" /></td></tr>
</tbody></table>
</div>
<p>For all target types, the <code>mosdepth</code> will produce various charts for
the whole genome, exonic, and coding sequence ranges. The most important chart
to examine is the &quot;cumulative coverage distribution&quot; chart, which shows the
burndown chart of what percentage of the specified region is covered by a
particular read depth. In this way, you can assess how much of the specified
region is covered by a particular read depth.</p>
<p>When interpreting this chart, the most important thing to examine is (a) the
chart's behavior with respect to the sequencing modality you are reviewing and
(b) the relative similarities of the plots within a cohort. For (a), you can
expect that</p>
<ul>
<li>Whole genome data will show a significant cliff for which all places in the
genome meet a particular coverage metric. There should be a pronounced
downward curve that signifies the cliff of &quot;normal&quot; sites across the genome
being separated from highly-covered sites of the genome.</li>
<li>Whole exome data will quickly drop off from left to right, as only about 3% of
the genome is targetted (the exonic regions). For those regions, however, you
can expect upwards of 1000x coverage.</li>
<li>For RNA-Seq, this plot is not as useful. As such, you can ignore it for this
sequencing modality.</li>
</ul>
<h3 id="ngsderive"><a class="header" href="#ngsderive">ngsderive</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td><code>ngsderive instrument</code></td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted sequencing instrument for this file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td><code>ngsderive readlen</code></td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted read length for the file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td><code>ngsderive encoding</code></td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted FASTQ quality score encoding scheme for the file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td><code>ngsderive strandedness</code></td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted strandedness of the RNA-Seq protocol for the file.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td><code>ngsderive junction-annotation</code></td><td><a href="https://github.com/stjudecloud/ngsderive/">ngsderive</a></td><td>The predicted breakdown of novel, partially novel, and known splice junctions.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
</tbody></table>
</div>
<p><code>ngsderive</code> is a tool written by the St. Jude Cloud team to provide forensic
analysis techniques for evaluating next-generation sequencing data. The
following subcommands are particularly useful in the context of these QC
reports:</p>
<ul>
<li><strong><code>ngsderive instrument</code>.</strong> Gives an estimate of which sequencing instrument
was used for a particular file. You should check to ensure the result you
receive from this subcommand matches your expectations of the data.</li>
<li><strong><code>ngsderive readlen</code>.</strong> Gives an estimate of the pre-trimmed read length. For
files with significant adapter trimming, this may not provide conclusive
results (giving a result of <code>-1</code>). You should check to ensure the result you
receive from this subcommand matches your expectations of the data.</li>
<li><strong><code>ngsderive encoding</code>.</strong> Guesses which <a href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">FASTQ
encoding</a> was used for
the original FASTQ file. For most modern data, this should be <code>Sanger/Illumina 1.8</code>.</li>
<li><strong><code>ngsderive strandedness</code>.</strong> For RNA-Seq data, derives the strandedness of
the sequencing protocol. This is useful to understand the effectiveness of the
assay and also how strong downstream results will be (if you believe your data
is <code>Reverse-stranded</code>, but your data appears to be somewhere between 
<code>Reverse-stranded</code> and <code>Unstranded</code>, then you may have issues with your
analysis).</li>
<li><strong><code>ngsderive junction-annotation</code>.</strong> For RNA-Seq data, reports the number of
novel junctions, partially novel junctions, and known junctions. This is
incredibly helpful when comparing with other samples within the same cohort.
Unless you have reason to suspect otherwise, you should expect the majority of
junctions to be known, followed by novel, and then partially novel junctions.
An abundance of novel junctions can indicate issues with the sample.</li>
</ul>
<p>For each of these, we recommend you view the graphs provided to pull out any
outliers in the data. In particular, the quantitative measures like percent
strandedness and junction saturation can give you strong indications if your
library preparation acted as you expected. For any outliers, you should contact
your lab to report the issue and determine if they noted any quality control
issues from their side.</p>
<h3 id="qualimap-rna-seq"><a class="header" href="#qualimap-rna-seq">Qualimap RNA-Seq</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>Genomic Origin of Reads</td><td><a href="http://qualimap.bioinfo.cipf.es/doc_html/command_line.html">qualimap</a></td><td>The proportion of intronic, exonic, and intergenic reads from RNA-Seq data.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Gene Coverage Profile</td><td><a href="http://qualimap.bioinfo.cipf.es/doc_html/command_line.html">qualimap</a></td><td>The distribution of read coverage across gene transcripts.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
</tbody></table>
</div>
<p>Qualimap RNA-Seq provides specific tools evaluating more straightforward metrics
from RNA-Seq data. In our QC pipeline, we use two graphs: (a) the genomic origin
of reads graph and (b) the gene coverage profile graph.</p>
<ul>
<li>The <strong>genomic origin of reads</strong> graph shows the breakdown of reads from
intronic, intergenic, and exonic regions. Particularly for PolyA selected
data, you should expect to see exonic regions dominate the distribution. For
Total or ribosomal depletion methods, you can expect to see more intronic
regions (as pre-mRNA are typically captured with these library preparation
protocols). Note that a high proportion of intergenic reads is often a sign
that your library selection has gone awry.</li>
<li>The <strong>gene coverage profile</strong> graph demonstrates the coverage across different
positions of gene transcripts. You should expect to see a gentle slope with a
peak towards the middle of each gene. With respect to the ends, the 3' end of
the gene typically drops off more quickly than the 5' end of the gene. If you
see irregular coverage of gene transcripts, such as erradict patterns in
coverage, this can indicate some bias or selection in the fragment selection
of the sequencing experiment.</li>
</ul>
<h3 id="picard"><a class="header" href="#picard">Picard</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>Alignment Summary</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Number of aligned versus unaligned reads for a given file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Mean Read Length Distribution</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Mean read length of the file.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>GC Coverage Bias</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Picard's interpretation of a GC bias distribution.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>Insert Size Distribution</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>The distribution of computed insert sizes for the file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Deduplication stats</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Overview of duplication from reads within the file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/warn-orange" alt="warn" /></td></tr>
<tr><td>Base quality distribution</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>The count of bases with each quality score.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>SAM/BAM validation</td><td><a href="https://broadinstitute.github.io/picard/">picard</a></td><td>Validation of the SAM/BAM files.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
</tbody></table>
</div>
<p><a href="https://broadinstitute.github.io/picard/">Picard</a> provides a variety of important tools for our QC workflow, including:</p>
<ul>
<li>An <strong>alignment summary</strong> chart, which illustrates the number of reads and the
relative breakdown of mapped versus unmapped reads. This is useful when
determining which files are over/under sequenced within a particular cohort
and for visually identifying samples with lower mapping rates (an indication
of a problem).</li>
<li>The <strong>mean read length</strong> chart, which we typically ignore in favor of
<code>ngsderive</code>'s <code>readlen</code> command (note that <code>ngsderive</code> will give a result of
<code>-1</code> if a result cannot be determined, whereas Picard gives you the median
read length).</li>
<li>The <strong>GC coverage bias</strong> chart, which we typically ignore in favor of the &quot;Per
Sequence GC Bias&quot; plot from FastQC.</li>
<li>The <strong>insert size distribution</strong> chart, which is <em>incredibly</em> helpful when
determining problems in your data. For example, if you have bacteria
contamination in your data, the bacterial reads will map with low quality
using <code>bwa mem</code> (and default parameters). This will result in a bimodal
distrbution of the insert size estimate, which is highly indicate of some
issue in the data. For these plots, (a) ensure that all of the insert size
distributions match each other within a cohort and (b) watch out for bimodal
distributions with a small (or large) peak around ~20bp, which is an
indication of contamination.</li>
<li>The <strong>deduplication stats</strong> chart, which give an indication of duplication within
the sample (note that, for RNA-Seq data, if this plot is shown, biological
duplicates cannot be ruled out!).</li>
<li>The <strong>base quality distribution</strong> chart, which we do not use and instead use
FastQC's &quot;Per Sequence Quality Scores&quot; plot.</li>
<li>The <strong>file validation</strong> chart, all of which should be green.</li>
</ul>
<h3 id="samtools"><a class="header" href="#samtools">Samtools</a></h3>
<p>Samtools also outputs various distributions as a beeswarm plot. In our
experience, these are not as useful expect in the case of examining a particular
sample to see where it lies for all metrics at once‚Äîotherwise, the summary
statistics in the &quot;General Statistics&quot; section at the top do just as well.</p>
<h3 id="kraken"><a class="header" href="#kraken">Kraken</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>Top Taxa</td><td><a href="https://github.com/DerrickWood/kraken2">kraken</a></td><td>Proportion of reads which are assigned to the indicated species.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
</tbody></table>
</div>
<p><a href="https://github.com/DerrickWood/kraken2">Kraken</a> is useful for determining the origin of reads within your sample of
interest. Since we mainly are sequencing human subjects, we typically use kraken
to give an indication of the level of contamination from various species within
our data. You should look to ensure that the vast majority (80% or higher) of
your sample originates from humans (<em>Homo sapiens</em>). If you are sequencing a
xenograft, pay close attention to the proportion of mouse (<em>Mus musculus</em>) reads
in the sample.</p>
<h3 id="fastqc"><a class="header" href="#fastqc">FastQC</a></h3>
<div class="table-wrapper"><table><thead><tr><th>Name</th><th>Tool</th><th>Description</th><th>WGS</th><th>WXS</th><th>RNA-Seq</th></tr></thead><tbody>
<tr><td>Sequencing Counts</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>Unique and duplicate sequences in the file.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>Sequencing Quality Histogram</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>Mean quality for every position in a read.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Per Sequence Quality Scores</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The number of reads with the indicated average quality score.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Per Base Sequence Content</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The bias towards a particular nucleotide at a given position of a read.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Per Sequence GC Content</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The GC content distribution.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Per Base N Content</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The percentile of base calls at each position which contained an N.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Sequence Length Distribution</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The distribution of read lengths in the file.</td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td><td><img src="https://img.shields.io/badge/no-red" alt="no" /></td></tr>
<tr><td>Sequence Duplication Levels</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The relative level of duplication for each sequence.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Overrepresented Sequences</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The sequences which may be overrepresented in your file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Adapter Content</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>The presence or absence of adapter contamination in your file.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
<tr><td>Status Checks</td><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></td><td>An overview of the pass/warn/fail status for all FastQC checks.</td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td><td><img src="https://img.shields.io/badge/yes-brightgreen" alt="yes" /></td></tr>
</tbody></table>
</div>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> rounds out the list of important tools to interrogate the quality of
the sequencing data itself. We leverage many of the plots produced by FastQC,
using it as the authority on most sequencing quality metrics, including:</p>
<ul>
<li>The <strong>sequencing counts</strong> chart, which we typically ignore in favor of
<a href="https://broadinstitute.github.io/picard/">picard</a>'s mark duplicates plot.</li>
<li>The <strong>sequence quality histogram</strong>, which gives an indication of the
sequencing quality across reads. Look for anything that fails FastQC's quite
stringent check to see if that correlates with any other failing metrics.</li>
<li>The <strong>per sequence quality scores</strong> chart, which shows the mean sequence
quality scores across the sample (peaks for lower scores in the orange or red
areas of the graph are cause for concern).</li>
<li>The <strong>per base sequence content</strong>, which shows is there are any biases in what
nucleotides are showing up at particular locations in the reads. For whole
genome and whole exome data, there should be relatively few biases. However,
for RNA-Seq, you will see a bias at the 5' region of the reads.</li>
<li>The <strong>per sequence GC content</strong> chart, which is highly informative of
contamination in whole genome and whole exome sequencing data. In those two
sequencing modalities, you should check to ensure that all of the
distributions are tightly following one another‚Äîany derivation from the
expected distribution is a sign of contamination. Note that sometimes tools
like FastQC plot an &quot;expected&quot; distribution of the GC content; this is usually
the expectation of WGS data, but not WXS data, which has a different expected
distribution. For RNA-Seq, this graph is generally not used except to examine
extreme biases at particular positions.</li>
<li>The <strong>per base N content</strong> chart, which should generally report very low
levels of Ns in modern data.</li>
<li>The <strong>sequence length distribution</strong> chart, which we typically ignore in favor
of <code>ngsderive</code>'s <code>readlen</code> plot.</li>
<li>The <strong>sequence duplication level</strong> chart, which gives an indication of the
percentage of sequences that are duplicated and by what factor (though this is
a graph that we find, often, even good data fails FastQC's stringent test).</li>
<li>The <strong>overrepresented sequences</strong> chart, which gives an indication if you're
sequencing the same sequence over and over.</li>
<li>The <strong>adapter content</strong> chart, which we use to ensure there are minimal to no
signs of adapters in the data we receive.</li>
<li>And finally, the <strong>status checks</strong> view, which gives you an overview of which
samples passed/warned/failed different checks. Generally, this is a good
overview of the quality of the data, though we find that many of the charts
are often marked as &quot;failed&quot; for good data (in particular, the &quot;Per Base
Sequence Content&quot; [for RNA-Seq], the &quot;Per Sequence GC Content&quot;, and the
&quot;Sequence Length Distribution&quot; graphs).</li>
</ul>
<h2 id="specification-1"><a class="header" href="#specification-1">Specification</a></h2>
<p>These are generic instructions for running each of the tools in our pipeline. We run our pipeline as a <a href="https://github.com/stjudecloud/workflows/blob/master/workflows/qc/quality-check-standard.wdl">WDL workflow</a>. We have supplied examples of the commands used for each package. For the typical memory requirements of each command, please see our <a href="https://github.com/stjudecloud/workflows">WDL repository</a>.</p>
<h3 id="dependencies-1"><a class="header" href="#dependencies-1">Dependencies</a></h3>
<p>We presume Anaconda is available and installed. If not, please follow the link to <a href="https://www.anaconda.com/">Anaconda</a> first.</p>
<pre><code class="language-bash">conda create --name bio-qc \
    --channel bioconda \
    --channel conda-forge \
    fastqc==0.11.8 \
    picard==2.20.2  \
    qualimap==2.2.2c \
    samtools==1.9 \
    fastq-screen==0.13.0 \
    ngsderive==2.2.0 \
    multiqc==1.10.1 \
    -y

conda activate bio-qc
</code></pre>
<p>For linting created fastqs, <code>fqlib</code> must be installed. See installation instructions <a href="https://github.com/stjude/fqlib">here</a>.</p>
<h3 id="workflow-1"><a class="header" href="#workflow-1">Workflow</a></h3>
<p>The workflow specification is as follows. Note that arguments that are not integral to the command (such as output directories) or vary between compute environments (such as memory thresholds or number of threads) are not included.</p>
<ol>
<li>
<p>Compute the md5 checksum of the file.</p>
<pre><code class="language-bash">md5sum $BAM
</code></pre>
</li>
<li>
<p>Use Picard's <code>ValidateSamFile</code> tool to ensure the inner contents of the BAM file are well-formed.</p>
<pre><code class="language-bash">picard ValidateSamFile \
    I=$BAM             \  # specify bam file
    MODE=SUMMARY          # concise output
</code></pre>
</li>
<li>
<p>Run <code>samtools flagstat</code> to gather general statistics such as alignment percentage.</p>
<pre><code class="language-bash">samtools flagstat $BAM
</code></pre>
</li>
<li>
<p>Run <code>fastqc</code> to collect sequencing and library-related statistics. These are only for informational purposes -- as stated above, we typically do not remove samples based on this information (with rare exception), as the sequencing-related QC work was done upstream in the genomics lab.</p>
<pre><code class="language-bash">fastqc $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive instrument</code> to infer sequencing instrument.</p>
<pre><code class="language-bash">ngsderive instrument $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive readlen</code> to infer read lengths.</p>
<pre><code class="language-bash">ngsderive readlen $BAM
</code></pre>
</li>
<li>
<p>Run <code>ngsderive encoding</code> to infer PHRED score encoding.</p>
<pre><code class="language-bash">ngsderive encoding \
    -n -1          \  # parse the entire file
    $BAM
</code></pre>
</li>
<li>
<p>Run <code>qualimap bamqc</code> to gather more in-depth statistics about read stats, coverage, mapping quality, mismatches, etc.</p>
<pre><code class="language-bash">qualimap bamqc -bam $BAM \  # bam filename
    -nt $NUM_THREADS     \  # threads requested
    -nw 400                 # number of windows
</code></pre>
</li>
<li>
<p>If WGS or WES data, run <code>fastq_screen</code>. For performance, we subsample the input BAM using <code>samtools view -s $computed_fraction</code> before running it through <code>picard SamToFastq</code>. The resulting fastqs are validated with <code>fq lint</code> provided by <code>fqlib</code>.</p>
<pre><code class="language-bash">cat $fastq_1 $fastq_2 &gt; $combined_fastq
fastq_screen          \
    --aligner bowtie2 \
    $combined_fastq
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>ngsderive strandedness</code> to determine a backwards-computed strandedness of the RNA-Seq experiment.</p>
<pre><code class="language-bash">ngsderive strandedness $BAM
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>ngsderive junction-annotation</code> to calculate the number of known, novel, and partial-novel junctions.</p>
<pre><code class="language-bash">ngsderive junction-annotation $BAM
</code></pre>
</li>
<li>
<p>If RNA-Seq data, run <code>qualimap rnaseq</code> to gather QC statistics that are tailored for RNA-Seq files.</p>
<pre><code class="language-bash">qualimap rnaseq --java-mem-size=$MEM_SIZE \  # memory
   -bam $BAM                              \  # bam filename
   [-pe]                                     # specify paired end if paired end
</code></pre>
</li>
<li>
<p>Combine all of the above metrics using <code>multiqc</code>.</p>
<pre><code class="language-bash">multiqc . # recurse all files in '.'
</code></pre>
</li>
</ol>
<div style="break-before: page; page-break-before: always;"></div><h1 id="table-of-contents-2"><a class="header" href="#table-of-contents-2">Table of Contents <!-- omit in toc --></a></h1>
<ul>
<li><a href="0002-scRNA-Seq-workflow.html#introduction">Introduction</a></li>
<li><a href="0002-scRNA-Seq-workflow.html#motivation">Motivation</a></li>
<li><a href="0002-scRNA-Seq-workflow.html#discussion">Discussion</a></li>
<li><a href="0002-scRNA-Seq-workflow.html#specification">Specification</a></li>
</ul>
<h1 id="introduction-2"><a class="header" href="#introduction-2">Introduction</a></h1>
<p>This RFC lays out the specification for the scRNA-Seq mapping pipeline.</p>
<h1 id="motivation-2"><a class="header" href="#motivation-2">Motivation</a></h1>
<p>To provide the  community access to data from scRNA-Seq experiments performed at St. Jude Children's Research Hospital, we propose the following data harmonization pipeline. The goal of this pipeline is to provide harmonized alignments for scRNA-Seq data. For this pipeline, we will make no recommendations on downstream analysis, focusing instead on harmonizing the underlying sequencing data and leaving analysis decisions to the user.</p>
<h1 id="discussion-2"><a class="header" href="#discussion-2">Discussion</a></h1>
<p>The development of bulk RNA-Seq enabled efficient profiling of transcripts in a sample. The ability to interrogate the transcriptome in an unbiased manner, largely replaced previous methods such as microarrays or RT-qPCR. A bulk RNA-Seq experiment sample is a mixture of cells. This method is useful for analyses such as comparing expression between tumor and normal samples. However, this limits the analysis to the average expression over a population of cells and is unable to capture the heterogeneity of gene expression across individual cells. Newer methods have enabled RNA-Seq to be applied to single cells. This enables new questions about cell-specific transcriptome changes. This methodology has been employed in large projects such as the <a href="https://www.humancellatlas.org/">Human Cell Atlas</a> to capture the cellular diversity within organisms.</p>
<h2 id="scrna-seq-methodology"><a class="header" href="#scrna-seq-methodology">scRNA-Seq methodology</a></h2>
<p>Generally, scRNA-Seq differs somewhat from bulk RNA-Seq. Most approaches generate three key pieces of information: 1. cDNA fragment from the RNA transcript, 2. a cell barcode that identifies in which cell the RNA was expressed, and 3. a unique molecular identifier (UMI) to enable collapse of PCR duplicates.</p>
<p>There are several commerical options for performing scRNA-Seq. One such option is the 10x Chromium platform. In their approach using paired-end reads, the celluar barcode and UMI are found in read 1 and the transcript sequence is found in read 2. The 10x method is the most popular commercially available method, so handling data for that platform will be the priority.</p>
<p>The canonical tool for processing 10x Chromium scRNA-Seq data is their <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> tool. This tool uses STAR to perform alignment. It then uses the transcriptome annotation GTF to label reads as exonic, intronic, or intergenic. It then performs some additional refinement on the alignments to determine which reads are transcriptomic. These reads are then carried into UMI counting. The UMI counting step produces a raw gene by cell matrix of counts.</p>
<h2 id="processing-pipeline"><a class="header" href="#processing-pipeline">Processing Pipeline</a></h2>
<p>For processing of bulk RNA-Seq data, a common standard workflow is to align against the genome using <a href="https://github.com/alexdobin/STAR">STAR</a>. Quantification is then done using a package such as <a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html">HTSeq-count</a>. An example of a bulk RNA-Seq pipeline is the <a href="0001-rnaseq-workflow-v2.0.0.html">St. Jude Cloud RNA-Seq standard workflow v2.0.0</a>.</p>
<h3 id="aligner-choice"><a class="header" href="#aligner-choice">Aligner choice</a></h3>
<p>For consistency with other St. Jude Cloud datasets, we are not considering psuedoalignment methods such as <a href="https://www.kallistobus.tools/">kallisto</a> and <a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin</a>. This leaves the previously described Cell Ranger method and STARsolo as the primary candidates for use in St. Jude Cloud.</p>
<p><a href="https://doi.org/10.1093/gigascience/giac001">Bruning, et. al.</a> (2022) provide a useful comparison of commonly used scRNA-Seq methods. However due to our preference for an alignment-based method, we will choose between Cell Ranger and STARsolo. Cell Ranger is the most commonly used method for processing data generated from the 10x protocol. Downstream analysis tools have been developed to directly ingest data in the formats produced by Cell Ranger. The tool is an integrated system that performs alignment, feature counting, and quality control.</p>
<img src="../resources/0002-scRNA-Seq-workflow/aligner_comparison.jpeg" width="800">
<h3 id="quantification-choice"><a class="header" href="#quantification-choice">Quantification choice</a></h3>
<p>Quantification is a product of both the Cell Ranger and STARsolo packages. Quantification could also be accomplished using a pseudoalignment method, such as Kallisto or Alevin, if there was no desire to share the raw data in BAM format. However that would be a departure from the normal St. Jude Cloud methodology, so that will not be discussed here. Therefore, there is no additional quantification choice to be made beyond that of the aligner package.</p>
<p>Using either STARsolo or Cell Ranger for analysis requires selection of reference files. For the genome, we will continue using GRCh38_no_alt. The transcript annotation GTF is more complicated. 10x <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest">provides</a> several versions of the annotation files used in Cell Ranger. For simplicity, we will use the Cell Ranger reference files version 2020-A which incorporates GENCODE v32.</p>
<h1 id="specification-2"><a class="header" href="#specification-2">Specification</a></h1>
<h2 id="metadata"><a class="header" href="#metadata">Metadata</a></h2>
<h3 id="sample-based"><a class="header" href="#sample-based">Sample-based</a></h3>
<ol>
<li>Library Prep Kit - e.g. Illumina TruSeq RNA Library Prep Kit V2</li>
<li>Library Selection Protocol - e.g. Total, Poly-A</li>
<li>Sequencing Machine - e.g. Illumina HiSeq 2000, Illumina NovaSeq 6000</li>
<li>Library Layout - Single-End, Paired-End</li>
<li>RNA-Seq Strandedness - Unstranded, Stranded-Reverse, Stranded-Forward</li>
<li>Read Length - Number of bp per read</li>
<li>Tissue Preservative - e.g. FFPE, Fresh/Frozen</li>
<li>Cell Isolation Protocol - e.g. Droplet, Microfluidics</li>
<li>Flowcell design - Multiplexed, split across flowcells, etc.</li>
</ol>
<h3 id="cell-based"><a class="header" href="#cell-based">Cell-based</a></h3>
<p>TODO</p>
<h2 id="dependencies-2"><a class="header" href="#dependencies-2">Dependencies</a></h2>
<p>If you'd like the full <code>conda</code> environment, you can install it using the following command. Obviously, you'll need to install <a href="https://www.anaconda.com/">anaconda</a> first.</p>
<pre><code class="language-bash">conda create -n scrnaseq-mapping \
    -c conda-forge \
    -c bioconda \
    picard==2.20.2 \
    samtools==1.9 \
    ngsderive==2.2.0 \
    tabix==1.11 \
    -y
</code></pre>
<p>Additionally, you will want to install our <code>fqlib</code> library to check that FastQ files are properly paired and have no duplicate read names. Installation of the <a href="https://rustup.rs/">Rust</a> programming language is required.</p>
<pre><code class="language-bash">cargo install --git https://github.com/stjude/fqlib.git --tag v0.8.0
</code></pre>
<p>Additionally, you will need to install <code>Cell Ranger</code> from 10x genomics to perform the alignment and feature calling.</p>
<pre><code class="language-bash">curl -o cellranger-7.0.0.tar.gz &quot;https://cf.10xgenomics.com/releases/cell-exp/cellranger-7.0.0.tar.gz?Expires=1659064381&amp;Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9jZi4xMHhnZW5vbWljcy5jb20vcmVsZWFzZXMvY2VsbC1leHAvY2VsbHJhbmdlci03LjAuMC50YXIuZ3oiLCJDb25kaXRpb24iOnsiRGF0ZUxlc3NUaGFuIjp7IkFXUzpFcG9jaFRpbWUiOjE2NTkwNjQzODF9fX1dfQ__&amp;Signature=SGKFTznrgspKBPQ2oRZ5C9KazslNWsTkxtkGnnaaPtrOR2iihhvkJ5frOlT5ahkABzBlHf~8EjLqcJ6HofbiQ4McbrHsmAnRggfv2QK0WScF40qDE3kGm~Z57VumO5homkdJPEf9r5DlbMzlArE5-UBH91F0rMMribGjwABXJCjUsAuet0klbUg~~SzCxoNMfTvo3Nn7Mxv7ls51LQf72riNit9zne6WsHynIY7YBHaquVftTi-C6bMZw5A3NoekyA7LBTZwc6sFjrsFrf3s3BpYKeRkBtPdkDMljME9JLDo9wXM7Lf1SfTj1vtNVUDoNhMnTFplDNzyBaDEDm7GVg__&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&quot;

echo '30855cb96a097c9cab6b02bdb520423f  cellranger-7.0.0.tar.gz' &gt; cellranger-7.0.0.tar.gz.md5

md5sum -c cellranger-7.0.0.tar.gz.md5

tar -xzvf cellranger-7.0.0.tar.gz

export PATH=$(pwd)/cellranger-7.0.0:$PATH
</code></pre>
<h2 id="reference-files-1"><a class="header" href="#reference-files-1">Reference files</a></h2>
<p>The following reference files are used as the basis of the scRNA-Seq Workflow v1.0.0:</p>
<ul>
<li>
<p>We will use the 10x provided reference files version 2020-A. You can get the file by running the following commands:</p>
<pre><code class="language-bash">wget https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz -O 10x-GRCh38-2020-A.tar.gz

echo &quot;dfd654de39bff23917471e7fcc7a00cd  10x-GRCh38-2020-A.tar.gz&quot; &gt; 10x-GRCh38-2020-A.tar.gz.md5
md5sum -c 10x-GRCh38-2020-A.tar.gz.md5
# &gt; 10x-GRCh38-2020-A.tar.gz: OK

tar -zxf 10x-GRCh38-2020-A.tar.gz
mv refdata-gex-GRCh38-2020-A 10x-GRCh38-2020-A
</code></pre>
</li>
</ul>
<h2 id="workflow-2"><a class="header" href="#workflow-2">Workflow</a></h2>
<p>Here are the resulting steps in the scRNA-Seq Workflow v1.0.0 pipeline. There might be slight alterations in the actual implementation, which can be found in <a href="https://github.com/stjudecloud/workflows/blob/master/workflows/scrnaseq/scrnaseq-standard.wdl">the St. Jude Cloud workflows repository</a>.</p>
<ol>
<li>
<p>Run <code>picard ValidateSam</code> on the incoming BAM to ensure that it is well-formed enough to convert back to FastQ.</p>
<pre><code class="language-bash">picard ValidateSamFile \
                  I=$INPUT_BAM \                     # Input BAM.
                  IGNORE=INVALID_PLATFORM_VALUE \    # Validations to ignore.
                  IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Split BAM file into multiple BAMs on the different read groups using <code>samtools split</code>. See <a href="http://www.htslib.org/doc/samtools.html">the samtools documentation</a> for more information.</p>
<pre><code class="language-bash">samtools split -u $UNACCOUNTED_BAM_NAME \ # Reads that do not belong to a read group or the read group is unrecognized go here.
               -f '%*_%!.%.'              # Format of output BAM file names.
</code></pre>
<p>If the BAM has unaccounted reads, those will need to be triaged and this step will need to be rerun.</p>
</li>
<li>
<p>Run Picard <code>SamToFastq</code> on each of the BAMs generated in the previous step.</p>
<pre><code class="language-bash">   picard SamToFastq \
          INPUT=$INPUT_BAM \
          FASTQ=$FASTQ_R1 \
          SECOND_END_FASTQ=$FASTQ_R2 \
          RE_REVERSE=true \
          VALIDATION_STRINGENCY=SILENT
</code></pre>
</li>
<li>
<p>Run <code>fq lint</code> on each of the FastQ pairs that was generated in the previous step as a sanity check. You can see the checks that the <code>fq</code> tool performs <a href="https://github.com/stjude/fqlib/blob/master/README.md#validators">here</a>.</p>
<pre><code class="language-bash">fq lint $FASTQ_R1 $FASTQ_R2 # Files for read 1 and read 2.
</code></pre>
</li>
<li>
<p>Run <code>Cell Ranger</code>.</p>
<pre><code class="language-bash">     cellranger count \
                --id=$OUTPUT_DIR_NAME \      # A unique run id and output folder name [a-zA-Z0-9_-]+
                --transcriptome=$REFERENCE \ # Path of folder containing 10x-compatible transcriptome reference
                --fastqs=$FASTQ_DIR \        # Path to input FASTQ data
                --sample=$FASTQ_PREFIX \     # Prefix of the filenames of FASTQs to select
                --localcores=$N_CORES \      # Maximum number of cores to use
                --localmem=$MEM_GB           # Maximum memory (GB) to use
</code></pre>
</li>
<li>
<p>Run <code>picard ValidateSamFile</code> on the aligned and sorted BAM file.</p>
<pre><code class="language-bash">picard ValidateSamFile \
               I=$CELL_RANGER_SORTED_BAM \     # Cell Ranger-aligned, coordinate-sorted BAM.
               IGNORE=INVALID_PLATFORM_VALUE \ # Validations to ignore.
               IGNORE=MISSING_PLATFORM_VALUE
</code></pre>
</li>
<li>
<p>Run <code>ngsderive</code>'s <code>strandedness</code> subcommand to confirm that the lab's information on strandedness reflects what was is computed. Manually triage any discrepancies. This is particularly useful for historical samples. Additionally,
if the value for strandedness isn't known at run time, we can use the inferred value (if reported).</p>
<pre><code class="language-bash">ngsderive strandedness $CELL_RANGER_SORTED_BAM \  # Cell Ranger-aligned, coordinate-sorted BAM.
                       -g $GENCODE_GTF_V32_GZ     # GENCODE v32 GTF (gzipped)
</code></pre>
</li>
</ol>
<div style="break-before: page; page-break-before: always;"></div><h1 id="drafts"><a class="header" href="#drafts">Drafts</a></h1>
<p>The following are <em>candidate</em> RFCs that are being rendered for easy review. They are <em>not</em> accepted St. Jude Cloud RFCs. For more information please see <a href="https://github.com/stjudecloud/rfcs/pulls">the associated pull request</a>.</p>
<ul>
<li><a href="https://stjudecloud.github.io/rfcs/branches/rfcs/chipseq_workflow">rfcs/chipseq_workflow</a></li>
<li><a href="https://stjudecloud.github.io/rfcs/branches/rfcs/qc-workflow">rfcs/qc-workflow</a></li>
<li><a href="https://stjudecloud.github.io/rfcs/branches/rfcs/structural-variation">rfcs/structural-variation</a></li>
</ul>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->

        <script>
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>

    </div>
    </body>
</html>
